diff --git a/coregrind/_Sys b/coregrind/_Sys
new file mode 100644
index 000000000..6c5ee1bb6
--- /dev/null
+++ b/coregrind/_Sys
@@ -0,0 +1,28 @@
+if &cp | set nocp | endif
+let s:cpo_save=&cpo
+set cpo&vim
+inoremap <C-U> u
+nmap Q gq
+xmap Q gq
+omap Q gq
+xmap gx <Plug>NetrwBrowseXVis
+nmap gx <Plug>NetrwBrowseX
+xnoremap <silent> <Plug>NetrwBrowseXVis :call netrw#BrowseXVis()
+nnoremap <silent> <Plug>NetrwBrowseX :call netrw#BrowseX(netrw#GX(),netrw#CheckIfRemote(netrw#GX()))
+inoremap  u
+let &cpo=s:cpo_save
+unlet s:cpo_save
+set background=dark
+set backspace=indent,eol,start
+set display=truncate
+set fileencodings=ucs-bom,utf-8,default,latin1
+set history=200
+set incsearch
+set mouse=a
+set nrformats=bin,hex
+set ruler
+set scrolloff=5
+set showcmd
+set ttimeout
+set ttimeoutlen=100
+" vim: set ft=vim :
