diff --git a/coregrind/m_aspacemgr/aspacemgr-linux.c b/coregrind/m_aspacemgr/aspacemgr-linux.c
index bae4f781a..bc7731da5 100644
--- a/coregrind/m_aspacemgr/aspacemgr-linux.c
+++ b/coregrind/m_aspacemgr/aspacemgr-linux.c
@@ -30,7 +30,8 @@
    The GNU General Public License is contained in the file COPYING.
 */
 
-#if defined(VGO_linux) || defined(VGO_darwin) || defined(VGO_solaris) || defined(VGO_freebsd)
+#if defined(VGO_linux) || defined(VGO_darwin) || defined(VGO_solaris) || defined(VGO_freebsd) \
+	|| defined(VGO_openedition)
 
 /* *************************************************************
    DO NOT INCLUDE ANY OTHER FILES HERE.
@@ -306,6 +307,8 @@ static Int      nsegments_used = 0;
 Addr VG_(clo_aspacem_minAddr)
 #if defined(VGO_linux)
    = (Addr) 0x04000000; // 64M
+#elif defined(VGO_openedition)
+   = (Addr) 0x04000000; //64M
 #elif defined(VGO_darwin)
 # if VG_WORDSIZE == 4
    = (Addr) 0x00001000;
@@ -882,7 +885,7 @@ static void sync_check_mapping_callback ( Addr addr, SizeT len, UInt prot,
 #endif
 
       /* Consider other reasons to not compare dev/inode */
-#if defined(VGO_linux)
+#if defined(VGO_linux) || defined(VGO_openedition)
       /* bproc does some godawful hack on /dev/zero at process
          migration, which changes the name of it, and its dev & ino */
       if (filename && 0==VG_(strcmp)(filename, "/dev/zero (deleted)"))
@@ -2754,6 +2757,17 @@ SysRes VG_(am_mmap_anon_float_valgrind)( SizeT length )
    }
 #endif
 
+#if defined(VGO_openedition) && !defined(ENABLE_INNER)
+    /* Doing the check only in linux not inner, as the below
+      check can fail when the kludge above has been used. */
+   if (sr_Res(sres) != advised) {
+      /* I don't think this can happen.  It means the kernel made a
+         fixed map succeed but not at the requested location.  Try to
+         repair the damage, then return saying the mapping failed. */
+      (void)ML_(am_do_munmap_NO_NOTIFY)( sr_Res(sres), length );
+      return VG_(mk_SysRes_Error)( VKI_EINVAL );
+   }
+#endif
    /* Ok, the mapping succeeded.  Now notify the interval map. */
    init_nsegment( &seg );
    seg.kind  = SkAnonV;
@@ -3346,7 +3360,7 @@ Bool VG_(am_relocate_nooverlap_client)( /*OUT*/Bool* need_discard,
 #endif // HAVE_MREMAP
 
 
-#if defined(VGO_linux)
+#if defined(VGO_linux) || defined(VGO_openedition)
 
 /*-----------------------------------------------------------------*/
 /*---                                                           ---*/
