diff --git a/configure.ac b/configure.ac
index cb1dc69d3..c6668bf2f 100755
--- a/configure.ac
+++ b/configure.ac
@@ -236,6 +236,11 @@ AC_MSG_CHECKING([for a supported CPU])
 # The names ppc64be or PPC64BE refer to only 64-bit systems that are Big
 # Endian.  Similarly, ppc64le or PPC64LE refer to only 64-bit systems that are
 # Little Endian.
+case "${host_os}" in
+     *openedition*)
+        host_cpu="s390x"
+        ;;
+esac
 
 VGCONF_PLATFORM_ARM_ARCH=
 
@@ -255,7 +260,6 @@ case "${host_cpu}" in
         AC_MSG_RESULT([ok (${host_cpu})])
         ARCH_MAX="ppc64be"
         ;;
-
      powerpc64le)
      # this only referrs to 64-bit Little Endian
         AC_MSG_RESULT([ok (${host_cpu})])
@@ -268,7 +272,7 @@ case "${host_cpu}" in
         ARCH_MAX="ppc32"
         ;;
 
-     s390x)
+     s390|s390x)
         AC_MSG_RESULT([ok (${host_cpu})])
         ARCH_MAX="s390x"
         ;;
@@ -518,6 +522,11 @@ case "${host_os}" in
         DEFAULT_SUPP="$srcdir/solaris12.supp ${DEFAULT_SUPP}"
         ;;
 
+     *openedition*)
+        AC_MSG_RESULT([ok (${host_os})])
+        VGCONF_OS="openedition"
+        ;;
+
      *) 
 	AC_MSG_RESULT([no (${host_os})])
 	AC_MSG_ERROR([Valgrind is operating system specific. Sorry.])
@@ -846,6 +855,21 @@ case "$ARCH_MAX-$VGCONF_OS" in
         valt_load_address_sec_inner="0xUNSET"
         AC_MSG_RESULT([ok (${ARCH_MAX}-${VGCONF_OS})])
         ;;
+     *openedition)
+       AC_MSG_RESULT([ok (s390x-openedition)])
+       VGCONF_ARCH_PRI="s390x"
+       VGCONF_ARCH_SEC=""
+       VGCONF_PLATFORM_PRI_CAPS="S390X_OPENEDITION"
+       VGCONF_PLATFORM_SEC_CAPS=""
+    # Use similar load addresses as s390x-linux
+    # To improve branch prediction hit rate we want to have
+    # the generated code close to valgrind (host) code
+       valt_load_address_pri_norml="0x800000000"
+       valt_load_address_pri_inner="0x810000000"
+       valt_load_address_sec_norml="0xUNSET"
+       valt_load_address_sec_inner="0xUNSET"
+       AC_MSG_RESULT([ok (${ARCH_MAX}-${VGCONF_OS})])
+       ;;
      mips32-linux) 
         VGCONF_ARCH_PRI="mips32"
         VGCONF_ARCH_SEC=""
@@ -1013,6 +1037,9 @@ AM_CONDITIONAL(VGCONF_PLATFORMS_INCLUDE_ARM64_LINUX,
 AM_CONDITIONAL(VGCONF_PLATFORMS_INCLUDE_S390X_LINUX,
                test x$VGCONF_PLATFORM_PRI_CAPS = xS390X_LINUX \
                  -o x$VGCONF_PLATFORM_SEC_CAPS = xS390X_LINUX)
+AM_CONDITIONAL(VGCONF_PLATFORMS_INCLUDE_S390X_OPENEDITION,
+               test x$VGCONF_PLATFORM_PRI_CAPS = xS390X_OPENEDITION \
+                 -o x$VGCONF_PLATFORM_SEC_CAPS = xS390X_OPENEDITION)
 AM_CONDITIONAL(VGCONF_PLATFORMS_INCLUDE_MIPS32_LINUX,
                test x$VGCONF_PLATFORM_PRI_CAPS = xMIPS32_LINUX \
                  -o x$VGCONF_PLATFORM_SEC_CAPS = xMIPS32_LINUX)
@@ -1061,6 +1088,8 @@ AM_CONDITIONAL(VGCONF_OS_IS_FREEBSD,
 AM_CONDITIONAL(VGCONF_OS_IS_DARWIN,
                test x$VGCONF_PLATFORM_PRI_CAPS = xX86_DARWIN \
                  -o x$VGCONF_PLATFORM_PRI_CAPS = xAMD64_DARWIN)
+AM_CONDITIONAL(VGCONF_OS_IS_OPENEDITION,
+               test x$VGCONF_PLATFORM_PRI_CAPS = xS390X_OPENEDITION)
 AM_CONDITIONAL(VGCONF_OS_IS_SOLARIS,
                test x$VGCONF_PLATFORM_PRI_CAPS = xX86_SOLARIS \
                  -o x$VGCONF_PLATFORM_PRI_CAPS = xAMD64_SOLARIS)
@@ -1228,6 +1257,10 @@ if test x$VGCONF_PLATFORM_PRI_CAPS = xX86_SOLARIS \
     GLIBC_VERSION="solaris"
 fi
 
+if test x$VGCONF_PLATFORM_PRI_CAPS = xS390X_OPENEDITION; then
+    GLIBC_VERSION="openedition"
+fi
+
 # GLIBC_VERSION is empty if a musl libc is used, so use the toolchain tuple
 # in this case.
 if test x$GLIBC_VERSION = x; then
@@ -1325,10 +1358,13 @@ case "${GLIBC_VERSION}" in
 	AC_DEFINE([MUSL_LIBC], 1, [Define to 1 if you're using Musl libc])
 	DEFAULT_SUPP="$srcdir/musl.supp ${DEFAULT_SUPP}"
 	;;
+     openedition)
+	AC_MSG_RESULT(openedition)
+	;;
      2.0|2.1|*)
 	AC_MSG_RESULT([unsupported version ${GLIBC_VERSION}])
 	AC_MSG_ERROR([Valgrind requires glibc version 2.2 or later, uClibc,])
-	AC_MSG_ERROR([musl libc, Darwin libc, Bionic libc or Solaris libc])
+	AC_MSG_ERROR([musl libc, Darwin libc, Bionic libc, Solaris libc or openedition])
 	;;
 esac
 
