diff --git a/VEX/priv/guest_s390_toIR.c b/VEX/priv/guest_s390_toIR.c
old mode 100644
new mode 100755
index 6da599645..d9d746c38
--- a/VEX/priv/guest_s390_toIR.c
+++ b/VEX/priv/guest_s390_toIR.c
@@ -8,7 +8,7 @@
    This file is part of Valgrind, a dynamic binary instrumentation
    framework.
 
-   Copyright IBM Corp. 2010-2024
+   Copyright IBM Corp. 2010-2021
 
    This program is free software; you can redistribute it and/or
    modify it under the terms of the GNU General Public License as
@@ -72,11 +72,7 @@ static DisResult *dis_res;
 static Bool sigill_diag;
 
 /* The last seen execute target instruction */
-enum { Invalid_execute_target = 1 };
-ULong last_execute_target = Invalid_execute_target;
-
-/* The guest address to be used as the base for relative addresses. */
-Addr64 guest_IA_rel_base;
+ULong last_execute_target;
 
 /* The possible outcomes of a decoding operation */
 typedef enum {
@@ -141,9 +137,9 @@ typedef enum {
 #define RIE_RRUUU_i3(insn) (((insn) >> 40) & 0xff)
 #define RIE_RRUUU_i4(insn) (((insn) >> 32) & 0xff)
 #define RIE_RRUUU_i5(insn) (((insn) >> 24) & 0xff)
-#define RIE_R0xU_r1(insn) (((insn) >> 52) & 0xf)
-#define RIE_R0xU_i2(insn) (((insn) >> 32) & 0xffff)
-#define RIE_R0xU_m3(insn) (((insn) >> 28) & 0xf)
+#define RIEv1_r1(insn) (((insn) >> 52) & 0xf)
+#define RIEv1_i2(insn) (((insn) >> 32) & 0xffff)
+#define RIEv1_m3(insn) (((insn) >> 28) & 0xf)
 #define RIE_RRPU_r1(insn) (((insn) >> 52) & 0xf)
 #define RIE_RRPU_r2(insn) (((insn) >> 48) & 0xf)
 #define RIE_RRPU_i4(insn) (((insn) >> 32) & 0xffff)
@@ -408,7 +404,7 @@ mkF64i(ULong value)
 static __inline__ Addr64
 addr_rel_long(UInt offset)
 {
-   return guest_IA_rel_base + ((Addr64)(Long)(Int)offset << 1);
+   return guest_IA_curr_instr + ((Addr64)(Long)(Int)offset << 1);
 }
 
 /* Return the 64-bit address with the given 16-bit "relative" offset from the
@@ -416,7 +412,7 @@ addr_rel_long(UInt offset)
 static __inline__ Addr64
 addr_relative(UShort offset)
 {
-   return guest_IA_rel_base + ((Addr64)(Long)(Short)offset << 1);
+   return guest_IA_curr_instr + ((Addr64)(Long)(Short)offset << 1);
 }
 
 /* Little helper function for my sanity. ITE = if-then-else */
@@ -556,27 +552,6 @@ system_call(IRExpr *sysno)
    dis_res->jk_StopHere = Ijk_Sys_syscall;
 }
 
-/* An extension */
-static void
-extension(ULong id, ULong variant)
-{
-   vassert(id < (1 << S390_EXT_ID_NBITS));
-   vassert(variant <= ~((ULong) 0) >> S390_EXT_ID_NBITS);
-
-   /* Store the extension ID in the pseudo register. */
-   ULong ext_id = id | (variant << S390_EXT_ID_NBITS);
-   stmt(IRStmt_Put(S390X_GUEST_OFFSET(guest_SYSNO), mkU64(ext_id)));
-
-   /* Store the current IA into guest_IP_AT_SYSCALL. */
-   stmt(IRStmt_Put(S390X_GUEST_OFFSET(guest_IP_AT_SYSCALL),
-                   mkU64(guest_IA_curr_instr)));
-
-   put_IA(mkaddr_expr(guest_IA_next_instr));
-
-   dis_res->whatNext    = Dis_StopHere;
-   dis_res->jk_StopHere = Ijk_Extension;
-}
-
 /* A side exit that branches back to the current insn if CONDITION is
    true. Does not set DisResult. */
 static void
@@ -2796,25 +2771,13 @@ s390_format_RIE_RRUUU(const HChar *(*irgen)(UChar r1, UChar r2, UChar i3,
 }
 
 static void
-s390_format_R0UU(const HChar *(*irgen)(UChar r1, UShort i2, UChar m3),
-                 UChar r1, UShort i2, UChar m3)
-{
-   const HChar *mnm = irgen(r1, i2, m3);
-
-   if (UNLIKELY(vex_traceflags & VEX_TRACE_FE))
-      s390_disasm(ENC4(XMNM, GPR, INT, CABM), S390_XMNM_CAB, mnm, m3, r1,
-                  i2, m3);
-}
-
-static void
-s390_format_R0IU(const HChar *(*irgen)(UChar r1, UShort i2, UChar m3),
-                 UChar r1, UShort i2, UChar m3)
+s390_format_RIEv1(const HChar *(*irgen)(UChar r1, UShort i2, UChar m3),
+                  UChar r1, UShort i2, UChar m3)
 {
    const HChar *mnm = irgen(r1, i2, m3);
 
    if (UNLIKELY(vex_traceflags & VEX_TRACE_FE))
-      s390_disasm(ENC4(XMNM, GPR, INT, CABM), S390_XMNM_CAB, mnm, m3, r1,
-                  (Int)(Short)i2, m3);
+      s390_disasm(ENC4(MNM, GPR, UINT, UINT), mnm, r1, i2, m3);
 }
 
 static void
@@ -2854,14 +2817,15 @@ s390_format_RIE_RUPI(const HChar *(*irgen)(UChar r1, UChar m3, UShort i4,
 }
 
 static void
-s390_format_RIE_RUPIX(const HChar *(*irgen)(UChar r1, UChar m3, UShort i2),
-                      UChar r1, UChar m3, UShort i2)
+s390_format_RIE_RUPIX(const HChar *(*irgen)(UChar r1, UChar m3, UShort i4,
+                                           UChar i2),
+                     UChar r1, UChar m3, UShort i4, UChar i2, Int xmnm_kind)
 {
-   const HChar *mnm = irgen(r1, m3, i2);
+   const HChar *mnm = irgen(r1, m3, i4, i2);
 
    if (UNLIKELY(vex_traceflags & VEX_TRACE_FE))
-      s390_disasm(ENC3(XMNM, GPR, INT), S390_XMNM_CLS, mnm, m3, r1,
-                  (Int)(Short)i2);
+      s390_disasm(ENC5(XMNM, GPR, INT, CABM, PCREL), xmnm_kind, mnm, m3, r1,
+                  (Int)(Char)i2, m3, (Int)(Short)i4);
 }
 
 static void
@@ -3119,10 +3083,10 @@ static void
 s390_format_RRF_U0RR(const HChar *(*irgen)(UChar m3, UChar r1, UChar r2),
                      UChar m3, UChar r1, UChar r2, Int xmnm_kind)
 {
-   const HChar *mnm = irgen(m3, r1, r2);
+   irgen(m3, r1, r2);
 
    if (UNLIKELY(vex_traceflags & VEX_TRACE_FE))
-      s390_disasm(ENC3(XMNM, GPR, GPR), xmnm_kind, mnm, m3, r1, r2);
+      s390_disasm(ENC3(XMNM, GPR, GPR), xmnm_kind, m3, r1, r2);
 }
 
 static void
@@ -3186,7 +3150,7 @@ s390_format_RRF_RURR(const HChar *(*irgen)(UChar, UChar, UChar, UChar),
    const HChar *mnm = irgen(r3, m4, r1, r2);
 
    if (UNLIKELY(vex_traceflags & VEX_TRACE_FE))
-      s390_disasm(ENC4(XMNM, GPR, GPR, GPR), S390_XMNM_CLS, mnm, m4, r1, r2, r3);
+      s390_disasm(ENC5(MNM, GPR, GPR, GPR, UINT), mnm, r1, r3, r2, m4);
 }
 
 static void
@@ -3345,28 +3309,10 @@ s390_format_RSY_RURD(const HChar *(*irgen)(UChar r1, UChar r3, IRTemp op2addr),
       s390_disasm(ENC4(MNM, GPR, UINT, SDXB), mnm, r1, r3, dh2, dl2, 0, b2);
 }
 
-static void
-s390_format_RSY_R0RD(const HChar *(*irgen)(UChar r1, UChar m3, IRTemp op2addr),
-                     UChar r1, UChar m3, UChar b2, UShort dl2, UChar dh2)
-{
-   const HChar *mnm;
-   IRTemp op2addr = newTemp(Ity_I64);
-   IRTemp d2 = newTemp(Ity_I64);
-
-   assign(d2, mkU64(((ULong)(Long)(Char)dh2 << 12) | ((ULong)dl2)));
-   assign(op2addr, binop(Iop_Add64, mkexpr(d2), b2 != 0 ? get_gpr_dw0(b2) :
-          mkU64(0)));
-
-   mnm = irgen(r1, m3, op2addr);
-
-   if (UNLIKELY(vex_traceflags & VEX_TRACE_FE))
-      s390_disasm(ENC4(XMNM, GPR, CABM, SDXB), S390_XMNM_CAB, mnm, m3, r1, m3,
-                  dh2, dl2, 0, b2);
-}
-
 static void
 s390_format_RSY_RDRM(const HChar *(*irgen)(UChar r1, IRTemp op2addr),
-                     UChar r1, UChar m3, UChar b2, UShort dl2, UChar dh2)
+                     UChar r1, UChar m3, UChar b2, UShort dl2, UChar dh2,
+                     Int xmnm_kind)
 {
    IRTemp op2addr = newTemp(Ity_I64);
    IRTemp d2 = newTemp(Ity_I64);
@@ -3377,13 +3323,12 @@ s390_format_RSY_RDRM(const HChar *(*irgen)(UChar r1, IRTemp op2addr),
    assign(op2addr, binop(Iop_Add64, mkexpr(d2), b2 != 0 ? get_gpr_dw0(b2) :
           mkU64(0)));
 
-   const HChar *mnm = irgen(r1, op2addr);
+   irgen(r1, op2addr);
 
    vassert(dis_res->whatNext == Dis_Continue);
 
    if (UNLIKELY(vex_traceflags & VEX_TRACE_FE))
-      s390_disasm(ENC3(XMNM, GPR, SDXB), S390_XMNM_CLS, mnm, m3, r1,
-                  dh2, dl2, 0, b2);
+      s390_disasm(ENC3(XMNM, GPR, SDXB), xmnm_kind, m3, r1, dh2, dl2, 0, b2);
 }
 
 static void
@@ -3562,18 +3507,6 @@ s390_format_S_RD(const HChar *(*irgen)(IRTemp op2addr),
       s390_disasm(ENC2(MNM, UDXB), mnm, d2, 0, b2);
 }
 
-static void
-s390_format_S_RD_raw(const HChar *(*irgen)(UChar b2, UShort d2),
-                     UChar b2, UShort d2)
-{
-   const HChar *mnm;
-
-   mnm = irgen(b2, d2);
-
-   if (UNLIKELY(vex_traceflags & VEX_TRACE_FE))
-      s390_disasm(ENC2(MNM, UDXB), mnm, d2, 0, b2);
-}
-
 static void
 s390_format_SI_URD(const HChar *(*irgen)(UChar i2, IRTemp op1addr),
                    UChar i2, UChar b1, UShort d1)
@@ -5408,14 +5341,14 @@ s390_irgen_BASR(UChar r1, UChar r2)
    IRTemp target = newTemp(Ity_I64);
 
    if (r2 == 0) {
-      put_gpr_dw0(r1, mkU64(guest_IA_next_instr));
+      put_gpr_dw0(r1, mkU64(guest_IA_curr_instr + 2ULL));
    } else {
       if (r1 != r2) {
-         put_gpr_dw0(r1, mkU64(guest_IA_next_instr));
+         put_gpr_dw0(r1, mkU64(guest_IA_curr_instr + 2ULL));
          call_function(get_gpr_dw0(r2));
       } else {
          assign(target, get_gpr_dw0(r2));
-         put_gpr_dw0(r1, mkU64(guest_IA_next_instr));
+         put_gpr_dw0(r1, mkU64(guest_IA_curr_instr + 2ULL));
          call_function(mkexpr(target));
       }
    }
@@ -5428,7 +5361,7 @@ s390_irgen_BAS(UChar r1, IRTemp op2addr)
 {
    IRTemp target = newTemp(Ity_I64);
 
-   put_gpr_dw0(r1, mkU64(guest_IA_next_instr));
+   put_gpr_dw0(r1, mkU64(guest_IA_curr_instr + 4ULL));
    assign(target, mkexpr(op2addr));
    call_function(mkexpr(target));
 
@@ -5598,7 +5531,7 @@ s390_irgen_BXLEG(UChar r1, UChar r3, IRTemp op2addr)
 static const HChar *
 s390_irgen_BRAS(UChar r1, UShort i2)
 {
-   put_gpr_dw0(r1, mkU64(guest_IA_next_instr));
+   put_gpr_dw0(r1, mkU64(guest_IA_curr_instr + 4ULL));
    call_function_and_chase(addr_relative(i2));
 
    return "bras";
@@ -5607,7 +5540,7 @@ s390_irgen_BRAS(UChar r1, UShort i2)
 static const HChar *
 s390_irgen_BRASL(UChar r1, UInt i2)
 {
-   put_gpr_dw0(r1, mkU64(guest_IA_next_instr));
+   put_gpr_dw0(r1, mkU64(guest_IA_curr_instr + 6ULL));
    call_function_and_chase(addr_rel_long(i2));
 
    return "brasl";
@@ -13202,24 +13135,27 @@ s390_irgen_TR_EX(IRTemp length, IRTemp start1, IRTemp start2)
 
 
 static void
-s390_irgen_EX_SS(UChar r, IRTemp addr2, IRTemp torun,
+s390_irgen_EX_SS(UChar r, IRTemp addr2,
                  void (*irgen)(IRTemp length, IRTemp start1, IRTemp start2),
                  UInt lensize)
 {
    IRTemp cond;
    IRDirty *d;
+   IRTemp torun;
    ULong ovl;
 
    IRTemp start1 = newTemp(Ity_I64);
    IRTemp start2 = newTemp(Ity_I64);
    IRTemp len = newTemp(lensize == 64 ? Ity_I64 : Ity_I32);
    cond = newTemp(Ity_I1);
+   torun = newTemp(Ity_I64);
 
+   assign(torun, load(Ity_I64, mkexpr(addr2)));
    /* Start with a check that the saved code is still correct */
    assign(cond, binop(Iop_CmpNE64, mkexpr(torun), mkU64(last_execute_target)));
    /* If not, save the new value */
    d = unsafeIRDirty_0_N (0, "s390x_dirtyhelper_EX", &s390x_dirtyhelper_EX,
-                          mkIRExprVec_2(mkexpr(torun), mkexpr(addr2)));
+                          mkIRExprVec_1(mkexpr(torun)));
    d->guard = mkexpr(cond);
    stmt(IRStmt_Dirty(d));
 
@@ -13238,37 +13174,21 @@ s390_irgen_EX_SS(UChar r, IRTemp addr2, IRTemp torun,
           r != 0 ? get_gpr_b7(r): mkU8(0), mkU8(SS_l(ovl)))));
    irgen(len, start1, start2);
 
-   last_execute_target = Invalid_execute_target;
+   last_execute_target = 0;
 }
 
 static const HChar *
 s390_irgen_EX(UChar r1, IRTemp addr2)
 {
-   IRTemp  insn0, unmodified_insn;
-   IRExpr* incr_addr;
-   insn0           = newTemp(Ity_I64);
-   unmodified_insn = newTemp(Ity_I64);
-   assign(insn0, unop(Iop_16Uto64, load(Ity_I16, mkexpr(addr2))));
-   incr_addr = binop(Iop_Add64, mkexpr(addr2), mkU64(2));
-   assign(
-      unmodified_insn,
-      binop(
-         Iop_Or64, binop(Iop_Shl64, mkexpr(insn0), mkU8(48)),
-         mkite(
-            binop(Iop_CmpLT64U, mkexpr(insn0), mkU64(0x4000)), mkU64(0),
-            mkite(binop(Iop_CmpLT64U, mkexpr(insn0), mkU64(0xc000)),
-                  binop(Iop_Shl64, unop(Iop_16Uto64, load(Ity_I16, incr_addr)),
-                        mkU8(32)),
-                  binop(Iop_Shl64, unop(Iop_32Uto64, load(Ity_I32, incr_addr)),
-                        mkU8(16))))));
-
-   if (last_execute_target == Invalid_execute_target) {
+   switch(last_execute_target & 0xff00000000000000ULL) {
+   case 0:
+   {
       /* no code information yet */
       IRDirty *d;
 
       /* so safe the code... */
       d = unsafeIRDirty_0_N (0, "s390x_dirtyhelper_EX", &s390x_dirtyhelper_EX,
-                             mkIRExprVec_2(mkexpr(unmodified_insn), mkexpr(addr2)));
+                             mkIRExprVec_1(load(Ity_I64, mkexpr(addr2))));
       stmt(IRStmt_Dirty(d));
       /* and restart */
       stmt(IRStmt_Put(S390X_GUEST_OFFSET(guest_CMSTART),
@@ -13280,43 +13200,42 @@ s390_irgen_EX(UChar r1, IRTemp addr2)
       put_IA(mkaddr_expr(guest_IA_next_instr));
       dis_res->whatNext = Dis_StopHere;
       dis_res->jk_StopHere = Ijk_InvalICache;
-      return "ex";
+      break;
    }
 
-   switch (last_execute_target & 0xff00000000000000ULL) {
    case 0xd200000000000000ULL:
       /* special case MVC */
-      s390_irgen_EX_SS(r1, addr2, unmodified_insn, s390_irgen_MVC_EX, 64);
+      s390_irgen_EX_SS(r1, addr2, s390_irgen_MVC_EX, 64);
       return "ex@mvc";
 
    case 0xd500000000000000ULL:
       /* special case CLC */
-      s390_irgen_EX_SS(r1, addr2, unmodified_insn, s390_irgen_CLC_EX, 64);
+      s390_irgen_EX_SS(r1, addr2, s390_irgen_CLC_EX, 64);
       return "ex@clc";
 
    case 0xd700000000000000ULL:
       /* special case XC */
-      s390_irgen_EX_SS(r1, addr2, unmodified_insn, s390_irgen_XC_EX, 32);
+      s390_irgen_EX_SS(r1, addr2, s390_irgen_XC_EX, 32);
       return "ex@xc";
 
    case 0xd600000000000000ULL:
       /* special case OC */
-      s390_irgen_EX_SS(r1, addr2, unmodified_insn, s390_irgen_OC_EX, 32);
+      s390_irgen_EX_SS(r1, addr2, s390_irgen_OC_EX, 32);
       return "ex@oc";
 
    case 0xd400000000000000ULL:
       /* special case NC */
-      s390_irgen_EX_SS(r1, addr2, unmodified_insn, s390_irgen_NC_EX, 32);
+      s390_irgen_EX_SS(r1, addr2, s390_irgen_NC_EX, 32);
       return "ex@nc";
 
    case 0xdc00000000000000ULL:
       /* special case TR */
-      s390_irgen_EX_SS(r1, addr2, unmodified_insn, s390_irgen_TR_EX, 64);
+      s390_irgen_EX_SS(r1, addr2, s390_irgen_TR_EX, 64);
       return "ex@tr";
 
    case 0xe800000000000000ULL:
       /* special case MVCIN */
-      s390_irgen_EX_SS(r1, addr2, unmodified_insn, s390_irgen_MVCIN_EX, 64);
+      s390_irgen_EX_SS(r1, addr2, s390_irgen_MVCIN_EX, 64);
       return "ex@mvcin";
 
    default:
@@ -13338,19 +13257,15 @@ s390_irgen_EX(UChar r1, IRTemp addr2)
       else
          assign(orperand, unop(Iop_8Uto64,get_gpr_b7(r1)));
       /* This code is going to be translated */
-      assign(torun, binop(Iop_Or64, mkexpr(unmodified_insn),
-                          binop(Iop_Shl64, mkexpr(orperand), mkU8(48))));
-
-      /* Start with a check that saved code is still correct. Compare the target
-       * address as well, since it may be relevant to relative addressing. */
-      assign(
-         cond,
-         binop(Iop_Or1,
-               binop(Iop_CmpNE64, mkexpr(torun), mkU64(last_execute_target)),
-               binop(Iop_CmpNE64, mkexpr(addr2), mkU64(guest_IA_rel_base))));
-      /* If not, save the new values */
+      assign(torun, binop(Iop_Or64, load(Ity_I64, mkexpr(addr2)),
+             binop(Iop_Shl64, mkexpr(orperand), mkU8(48))));
+
+      /* Start with a check that saved code is still correct */
+      assign(cond, binop(Iop_CmpNE64, mkexpr(torun),
+             mkU64(last_execute_target)));
+      /* If not, save the new value */
       d = unsafeIRDirty_0_N (0, "s390x_dirtyhelper_EX", &s390x_dirtyhelper_EX,
-                             mkIRExprVec_2(mkexpr(torun), mkexpr(addr2)));
+                             mkIRExprVec_1(mkexpr(torun)));
       d->guard = mkexpr(cond);
       stmt(IRStmt_Dirty(d));
 
@@ -13366,7 +13281,7 @@ s390_irgen_EX(UChar r1, IRTemp addr2)
       if (UNLIKELY(vex_traceflags & VEX_TRACE_FE))
          vex_printf("    which was executed by\n");
       /* dont make useless translations in the next execute */
-      last_execute_target = Invalid_execute_target;
+      last_execute_target = 0;
    }
    }
    return "ex";
@@ -13376,20 +13291,13 @@ static const HChar *
 s390_irgen_EXRL(UChar r1, UInt offset)
 {
    IRTemp addr = newTemp(Ity_I64);
-   Addr64 bytes_addr;
-   UChar *bytes;
+   Addr64 bytes_addr = addr_rel_long(offset);
+   UChar *bytes = (UChar *)(HWord)bytes_addr;
    /* we might save one round trip because we know the target */
-   if (last_execute_target == Invalid_execute_target) {
-      bytes_addr = addr_rel_long(offset);
-      bytes = (UChar *)(HWord)bytes_addr;
-      last_execute_target = ((ULong)bytes[0] << 56) | ((ULong)bytes[1] << 48);
-      if (bytes[0] >= 0x40)
-         last_execute_target |= ((ULong)bytes[2] << 40) | ((ULong)bytes[3] << 32);
-      if (bytes[0] >= 0xc0)
-         last_execute_target |= ((ULong)bytes[4] << 24) | ((ULong)bytes[5] << 16);
-      guest_IA_rel_base = bytes_addr;
-   } else
-      bytes_addr = guest_IA_rel_base;
+   if (!last_execute_target)
+      last_execute_target = ((ULong)bytes[0] << 56) | ((ULong)bytes[1] << 48) |
+                            ((ULong)bytes[2] << 40) | ((ULong)bytes[3] << 32) |
+                            ((ULong)bytes[4] << 24) | ((ULong)bytes[5] << 16);
    assign(addr, mkU64(bytes_addr));
    s390_irgen_EX(r1, addr);
    return "exrl";
@@ -14497,6 +14405,7 @@ static const HChar *
 s390_irgen_KEB(UChar r1, IRTemp op2addr)
 {
    return s390_irgen_CxB("keb", r1, op2addr, Ity_F32, Iop_CmpF32);
+   return "keb";
 }
 
 static const HChar *
@@ -15358,13 +15267,37 @@ s390_irgen_STCKE(IRTemp op2addr)
 }
 
 static const HChar *
-s390_irgen_STFLE(UChar b2, UShort d2)
+s390_irgen_STFLE(IRTemp op2addr)
 {
    if (! s390_host_has_stfle) {
       emulation_failure(EmFail_S390X_stfle);
       return "stfle";
    }
-   extension(S390_EXT_STFLE, b2 | (d2 << 8));
+
+   IRDirty *d;
+   IRTemp cc = newTemp(Ity_I64);
+
+   /* IRExpr_GSPTR() => Need to pass pointer to guest state to helper */
+   d = unsafeIRDirty_1_N(cc, 0, "s390x_dirtyhelper_STFLE",
+                         &s390x_dirtyhelper_STFLE,
+                         mkIRExprVec_2(IRExpr_GSPTR(), mkexpr(op2addr)));
+
+   d->nFxState = 1;
+   vex_bzero(&d->fxState, sizeof(d->fxState));
+
+   d->fxState[0].fx     = Ifx_Modify;  /* read then write */
+   d->fxState[0].offset = S390X_GUEST_OFFSET(guest_r0);
+   d->fxState[0].size   = sizeof(ULong);
+
+   d->mAddr = mkexpr(op2addr);
+   /* Pretend all double words are written */
+   d->mSize = S390_NUM_FACILITY_DW * sizeof(ULong);
+   d->mFx   = Ifx_Write;
+
+   stmt(IRStmt_Dirty(d));
+
+   s390_cc_set(cc);
+
    return "stfle";
 }
 
@@ -16656,11 +16589,20 @@ s390_irgen_VSCEG(UChar v1, IRTemp op2addr, UChar m3)
 static const HChar *
 s390_irgen_VPDI(UChar v1, UChar v2, UChar v3, UChar m4)
 {
-   /* These bits are reserved by specification */
-   s390_insn_assert("vpdi", (m4 & 2) == 0 && (m4 & 8) == 0);
+   /* Theese bits are reserved by specification */
+   vassert((m4 & 2) == 0);
+   vassert((m4 & 8) == 0);
+
+   if((m4 & 4) != 0)
+      put_vr_dw0(v1, get_vr_dw1(v2));
+   else
+      put_vr_dw0(v1, get_vr_dw0(v2));
+
+   if((m4 & 1) != 0)
+      put_vr_dw1(v1, get_vr_dw1(v3));
+   else
+      put_vr_dw1(v1, get_vr_dw0(v3));
 
-   put_vr_qw(v1, binop(Iop_64HLtoV128, m4 & 4 ? get_vr_dw1(v2) : get_vr_dw0(v2),
-                       m4 & 1 ? get_vr_dw1(v3) : get_vr_dw0(v3)));
    return "vpdi";
 }
 
@@ -17088,7 +17030,7 @@ s390_irgen_LOCFHR(UChar m3, UChar r1, UChar r2)
 }
 
 static const HChar *
-s390_irgen_LOCHHI(UChar r1, UChar m3, UShort i2)
+s390_irgen_LOCHHI(UChar r1, UChar m3, UShort i2, UChar unused)
 {
    next_insn_if(binop(Iop_CmpEQ32, s390_call_calculate_cond(m3), mkU32(0)));
    put_gpr_w0(r1, mkU32((UInt)(Int)(Short)i2));
@@ -17097,7 +17039,7 @@ s390_irgen_LOCHHI(UChar r1, UChar m3, UShort i2)
 }
 
 static const HChar *
-s390_irgen_LOCHI(UChar r1, UChar m3, UShort i2)
+s390_irgen_LOCHI(UChar r1, UChar m3, UShort i2, UChar unused)
 {
    next_insn_if(binop(Iop_CmpEQ32, s390_call_calculate_cond(m3), mkU32(0)));
    put_gpr_w1(r1, mkU32((UInt)(Int)(Short)i2));
@@ -17106,7 +17048,7 @@ s390_irgen_LOCHI(UChar r1, UChar m3, UShort i2)
 }
 
 static const HChar *
-s390_irgen_LOCGHI(UChar r1, UChar m3, UShort i2)
+s390_irgen_LOCGHI(UChar r1, UChar m3, UShort i2, UChar unused)
 {
    next_insn_if(binop(Iop_CmpEQ32, s390_call_calculate_cond(m3), mkU32(0)));
    put_gpr_dw0(r1, mkU64((ULong)(Long)(Short)i2));
@@ -17136,7 +17078,59 @@ s390_irgen_LCBB(UChar r1, IRTemp op2addr, UChar m3)
    return "lcbb";
 }
 
-/* Also known as "PRNO" */
+/* Regarding the use of 
+   // Dummy helper which is used to signal VEX library that memory was loaded
+   sha512_loadparam 
+     = unsafeIRDirty_0_N(0, "s390x_dirtyhelper_PPNO_sha512_load_param_block",
+                             &s390x_dirtyhelper_PPNO_sha512_load_param_block,
+                             mkIRExprVec_0());
+
+   in the following function (s390_irgen_PPNO).  This is a workaround to get
+   around the fact that IRDirty annotations cannot indicate two memory side
+   effects, which are unfortunately necessary here.  It will possibly lead to
+   losing undefinedness (undefinedness in some inputs might not be propagated
+   to the outputs as it shouod, in Memcheck).  The correct fix would be to
+   extend IRDirty to represent two memory side effects, but that's quite a bit
+   of work.
+
+   Here's a summary of what this insn does.
+
+   // getReg(RegisterNumber n) returns the value of GPR number 'n'
+
+   // reg1 and reg2 are even
+   void ppno(RegisterNumber reg1, RegisterNumber reg2) {
+
+       switch(getReg(0)) {
+       case 0x0:
+           // Query mode, ignore reg1 and reg2
+           // Write 16 bytes                    at  getReg(1)
+           break;
+
+       case 0x3:
+           // SHA-512 generate mode, ignore reg2
+
+           // Read 240 bytes                    at  getReg(1)
+           // Write getReg(reg1 + 1) bytes      at  getReg(reg1)
+           // Write some of 240 bytes starting  at  getReg(1)
+           break;
+
+       case 0x83:
+           // SHA-512 seed mode, ignore reg1
+
+           // Read some of 240 bytes starting  at  getReg(1)
+           // Read getReg(reg2 + 1) bytes      at  getReg(reg2)
+           // Write 240 bytes                  at  getReg(1)
+           break;
+
+       default:
+           // Specification exception, abort execution.
+       }
+   }
+*/
+/* Also known as "prno"
+   If you implement new functions please don't forget to update
+   "s390x_dirtyhelper_PPNO_query" function.
+ */
 static const HChar *
 s390_irgen_PPNO(UChar r1, UChar r2)
 {
@@ -17145,24 +17139,132 @@ s390_irgen_PPNO(UChar r1, UChar r2)
       return "ppno";
    }
 
-   /* Check for obvious specification exceptions */
-   s390_insn_assert("ppno", r1 % 2 == 0 && r2 % 2 == 0 && r1 != 0 && r2 != 0);
+   /* Theese conditions lead to specification exception */
+   vassert(r1 % 2 == 0);
+   vassert(r2 % 2 == 0);
+   vassert((r1 != 0) && (r2 != 0));
+
+   IRDirty *query, *sha512_gen, *sha512_seed, *sha512_loadparam;
+   IRTemp gpr1num = newTemp(Ity_I64);
+   IRTemp gpr2num = newTemp(Ity_I64);
+
+   IRTemp funcCode = newTemp(Ity_I8);
+   IRTemp is_query = newTemp(Ity_I1);
+   IRTemp is_sha512_gen = newTemp(Ity_I1);
+   IRTemp is_sha512_seed = newTemp(Ity_I1);
+   IRTemp is_sha512 = newTemp(Ity_I1);
+
+   assign(funcCode, unop(Iop_64to8, binop(Iop_And64, get_gpr_dw0(0),
+                                          mkU64(0xffULL))));
+   assign(gpr1num, mkU64(r1));
+   assign(gpr2num, mkU64(r2));
+
+   assign(is_query, binop(Iop_CmpEQ8, mkexpr(funcCode), mkU8(S390_PPNO_QUERY)));
+   assign(is_sha512_gen, binop(Iop_CmpEQ8, mkexpr(funcCode),
+                               mkU8(S390_PPNO_SHA512_GEN)));
+   assign(is_sha512_seed, binop(Iop_CmpEQ8, mkexpr(funcCode),
+                                mkU8(S390_PPNO_SHA512_SEED)));
+   assign(is_sha512, binop(Iop_CmpEQ8,
+                           mkU8(S390_PPNO_SHA512_GEN),
+                           binop(Iop_And8,
+                                 mkexpr(funcCode),
+                                 mkU8(S390_PPNO_SHA512_GEN)
+                                 )
+                           ));
+
+   query = unsafeIRDirty_0_N(0, "s390x_dirtyhelper_PPNO_query",
+                             &s390x_dirtyhelper_PPNO_query,
+                             mkIRExprVec_3(IRExpr_GSPTR(), mkexpr(gpr1num),
+                                           mkexpr(gpr2num)));
+   query->guard = mkexpr(is_query);
+   query->nFxState = 1;
+   vex_bzero(&query->fxState, sizeof(query->fxState));
+   query->fxState[0].fx     = Ifx_Read;
+   query->fxState[0].offset = S390X_GUEST_OFFSET(guest_r0);
+   query->fxState[0].size   = 2 * sizeof(ULong); /* gpr0 and gpr1 are read */
+   query->mAddr = get_gpr_dw0(1);
+   query->mSize = S390_PPNO_PARAM_BLOCK_SIZE_QUERY;
+   query->mFx   = Ifx_Write;
+
+   IRTemp gen_cc = newTemp(Ity_I64);
+   sha512_gen = unsafeIRDirty_1_N(gen_cc, 0, "s390x_dirtyhelper_PPNO_sha512",
+                             &s390x_dirtyhelper_PPNO_sha512,
+                             mkIRExprVec_3(IRExpr_GSPTR(), mkexpr(gpr1num),
+                                           mkexpr(gpr2num)));
+   sha512_gen->guard = mkexpr(is_sha512_gen);
+   sha512_gen->nFxState = 3;
+   vex_bzero(&sha512_gen->fxState, sizeof(sha512_gen->fxState));
+   sha512_gen->fxState[0].fx     = Ifx_Read;
+   sha512_gen->fxState[0].offset = S390X_GUEST_OFFSET(guest_r0);
+   sha512_gen->fxState[0].size   = 2 * sizeof(ULong); /* gpr0 and gpr1 are read */
+   sha512_gen->fxState[1].fx     = Ifx_Read;
+   sha512_gen->fxState[1].offset = S390X_GUEST_OFFSET(guest_r0) + r1 * sizeof(ULong);
+   sha512_gen->fxState[1].size   = sizeof(ULong);
+   sha512_gen->fxState[2].fx     = Ifx_Modify;
+   sha512_gen->fxState[2].offset = S390X_GUEST_OFFSET(guest_r0) + (r1 + 1) * sizeof(ULong);
+   sha512_gen->fxState[2].size   = sizeof(ULong);
+   sha512_gen->mAddr = get_gpr_dw0(r1);
+   sha512_gen->mSize = S390_PPNO_MAX_SIZE_SHA512_GEN;
+   sha512_gen->mFx   = Ifx_Write;
+
+   IRTemp unused = newTemp(Ity_I64);
+   sha512_seed = unsafeIRDirty_1_N(unused, 0, "s390x_dirtyhelper_PPNO_sha512",
+                             &s390x_dirtyhelper_PPNO_sha512,
+                             mkIRExprVec_3(IRExpr_GSPTR(), mkexpr(gpr1num),
+                                           mkexpr(gpr2num)));
+   sha512_seed->guard = mkexpr(is_sha512_seed);
+   sha512_seed->nFxState = 2;
+   vex_bzero(&sha512_seed->fxState, sizeof(sha512_seed->fxState));
+   sha512_seed->fxState[0].fx     = Ifx_Read;
+   sha512_seed->fxState[0].offset = S390X_GUEST_OFFSET(guest_r0);
+   sha512_seed->fxState[0].size   = 2 * sizeof(ULong); /* gpr0 and gpr1 are read */
+   sha512_seed->fxState[1].fx     = Ifx_Read;
+   sha512_seed->fxState[1].offset = S390X_GUEST_OFFSET(guest_r0) + r2 * sizeof(ULong);
+   sha512_seed->fxState[1].size   = 2 * sizeof(ULong); /* r2 and r2 + 1 are read */
+   sha512_seed->mAddr = get_gpr_dw0(r2);
+   sha512_seed->mSize = S390_PPNO_MAX_SIZE_SHA512_SEED;
+   sha512_seed->mFx   = Ifx_Write;
+
+   /* Dummy helper which is used to signal VEX library that memory was loaded */
+   sha512_loadparam =
+      unsafeIRDirty_0_N(0, "s390x_dirtyhelper_PPNO_sha512_load_param_block",
+                        &s390x_dirtyhelper_PPNO_sha512_load_param_block,
+                        mkIRExprVec_0());
+   sha512_loadparam->guard = mkexpr(is_sha512);
+   sha512_loadparam->nFxState = 0;
+   vex_bzero(&sha512_loadparam->fxState, sizeof(sha512_loadparam->fxState));
+   sha512_loadparam->mAddr = get_gpr_dw0(1);
+   sha512_loadparam->mSize = S390_PPNO_PARAM_BLOCK_SIZE_SHA512;
+   sha512_loadparam->mFx   = Ifx_Read;
+
+   IRDirty* sha512_saveparam =
+      unsafeIRDirty_0_N(0, "s390x_dirtyhelper_PPNO_sha512_save_param_block",
+                        &s390x_dirtyhelper_PPNO_sha512_load_param_block,
+                        mkIRExprVec_0());
+   sha512_saveparam->guard = mkexpr(is_sha512);
+   sha512_saveparam->nFxState = 0;
+   vex_bzero(&sha512_saveparam->fxState, sizeof(sha512_saveparam->fxState));
+   sha512_saveparam->mAddr = get_gpr_dw0(1);
+   sha512_saveparam->mSize = S390_PPNO_PARAM_BLOCK_SIZE_SHA512;
+   sha512_saveparam->mFx   = Ifx_Write;
+
+   stmt(IRStmt_Dirty(query));
+   stmt(IRStmt_Dirty(sha512_loadparam));
+   stmt(IRStmt_Dirty(sha512_gen));
+   stmt(IRStmt_Dirty(sha512_seed));
+   stmt(IRStmt_Dirty(sha512_saveparam));
 
-   extension(S390_EXT_PRNO, r1 | (r2 << 4));
-   return "ppno";
-}
-
-static const HChar *
-s390_irgen_DFLTCC(UChar r3, UChar r1, UChar r2)
-{
-   s390_insn_assert("dfltcc", s390_host_has_dflt);
+   IRTemp cc = newTemp(Ity_I64);
+   assign(cc,
+          mkite(mkexpr(is_sha512_gen),
+                mkexpr(gen_cc),
+                mkU64(0)
+               )
+         );
 
-   /* Check for obvious specification exceptions */
-   s390_insn_assert("dfltcc", r1 % 2 == 0 && r1 != 0 &&
-                    r2 % 2 == 0 && r2 != 0 && r3 >= 2);
+   s390_cc_thunk_fill(mkU64(S390_CC_OP_SET), mkexpr(cc), mkU64(0), mkU64(0));
 
-   extension(S390_EXT_DFLT, r1 | (r2 << 4) | (r3 << 8));
-   return "dfltcc";
+   return "ppno";
 }
 
 enum s390_VStrX {
@@ -19622,214 +19724,6 @@ s390_irgen_VSTEBRG(UChar v1, IRTemp op2addr, UChar m3)
    return "vstebrg";
 }
 
-static const HChar *
-s390_irgen_VCxx(const HChar *mnem, s390x_vec_op_details_t details,
-                UShort v2_offs, UShort v2_size)
-{
-   s390_insn_assert(mnem, s390_host_has_nnpa);
-
-   IRDirty* d = unsafeIRDirty_0_N(0, "s390x_dirtyhelper_vec_op",
-                                  &s390x_dirtyhelper_vec_op,
-                                  mkIRExprVec_2(IRExpr_GSPTR(),
-                                                mkU64(details.serialized)));
-   d->nFxState = 2;
-   vex_bzero(&d->fxState, sizeof(d->fxState));
-   d->fxState[0].fx = Ifx_Read;
-   d->fxState[0].offset = S390X_GUEST_OFFSET(guest_v0)
-      + details.v2 * sizeof(V128) + v2_offs;
-   d->fxState[0].size = v2_size;
-   d->fxState[1].fx = Ifx_Write;
-   d->fxState[1].offset = S390X_GUEST_OFFSET(guest_v0)
-      + details.v1 * sizeof(V128);
-   d->fxState[1].size = sizeof(V128);
-
-   stmt(IRStmt_Dirty(d));
-   return mnem;
-}
-
-static const HChar *
-s390_irgen_VCNF(UChar v1, UChar v2, UChar m3, UChar m4)
-{
-   s390x_vec_op_details_t details = { .serialized = 0ULL };
-   details.op = S390_VEC_OP_VCNF;
-   details.v1 = v1;
-   details.v2 = v2;
-   details.m3 = m3;
-   details.m4 = m4;
-   return s390_irgen_VCxx("vcnf", details, 0, sizeof(V128));
-}
-
-static const HChar *
-s390_irgen_VCLFNH(UChar v1, UChar v2, UChar m3, UChar m4)
-{
-   s390x_vec_op_details_t details = { .serialized = 0ULL };
-   details.op = S390_VEC_OP_VCLFNH;
-   details.v1 = v1;
-   details.v2 = v2;
-   details.m3 = m3;
-   details.m4 = m4;
-   return s390_irgen_VCxx("vclfnh", details, 0, sizeof(V128) / 2);
-}
-
-static const HChar *
-s390_irgen_VCFN(UChar v1, UChar v2, UChar m3, UChar m4)
-{
-   s390x_vec_op_details_t details = { .serialized = 0ULL };
-   details.op = S390_VEC_OP_VCFN;
-   details.v1 = v1;
-   details.v2 = v2;
-   details.m3 = m3;
-   details.m4 = m4;
-   return s390_irgen_VCxx("vcfn", details, 0, sizeof(V128));
-}
-
-static const HChar *
-s390_irgen_VCLFNL(UChar v1, UChar v2, UChar m3, UChar m4)
-{
-   s390x_vec_op_details_t details = { .serialized = 0ULL };
-   details.op = S390_VEC_OP_VCLFNL;
-   details.v1 = v1;
-   details.v2 = v2;
-   details.m3 = m3;
-   details.m4 = m4;
-   return s390_irgen_VCxx("vclfnl", details, sizeof(V128) / 2,
-                          sizeof(V128) / 2);
-}
-
-static const HChar *
-s390_irgen_VCRNF(UChar v1, UChar v2, UChar v3, UChar m4, UChar m5)
-{
-   s390_insn_assert("vcrnf", s390_host_has_nnpa);
-
-   s390x_vec_op_details_t details = { .serialized = 0ULL };
-   details.op = S390_VEC_OP_VCRNF;
-   details.v1 = v1;
-   details.v2 = v2;
-   details.v3 = v3;
-   details.m4 = m4;
-   details.m5 = m5;
-   details.m6 = 0;
-   IRDirty* d = unsafeIRDirty_0_N(0, "s390x_dirtyhelper_vec_op",
-                                  &s390x_dirtyhelper_vec_op,
-                                  mkIRExprVec_2(IRExpr_GSPTR(),
-                                                mkU64(details.serialized)));
-   d->nFxState = 3;
-   vex_bzero(&d->fxState, sizeof(d->fxState));
-   d->fxState[0].fx = Ifx_Read;
-   d->fxState[0].offset = S390X_GUEST_OFFSET(guest_v0) + v2 * sizeof(V128);
-   d->fxState[0].size = sizeof(V128);
-   d->fxState[1].fx = Ifx_Read;
-   d->fxState[1].offset = S390X_GUEST_OFFSET(guest_v0) + v3 * sizeof(V128);
-   d->fxState[1].size = sizeof(V128);
-   d->fxState[2].fx = Ifx_Write;
-   d->fxState[2].offset = S390X_GUEST_OFFSET(guest_v0) + v1 * sizeof(V128);
-   d->fxState[2].size = sizeof(V128);
-
-   stmt(IRStmt_Dirty(d));
-   return "vcrnf";
-}
-
-static const HChar *
-s390_irgen_NNPA(void)
-{
-   s390_insn_assert("nnpa", s390_host_has_nnpa);
-   extension(S390_EXT_NNPA, 0);
-   return "nnpa";
-}
-
-static const HChar *
-s390_irgen_KM(UChar r1, UChar r2)
-{
-   s390_insn_assert("km", r1 != 0 && r1 % 2 == 0 && r2 != 0 && r2 % 2 == 0);
-   extension(S390_EXT_KM, r1 | (r2 << 4));
-   return "km";
-}
-
-static const HChar *
-s390_irgen_KMC(UChar r1, UChar r2)
-{
-   s390_insn_assert("kmc", r1 != 0 && r1 % 2 == 0 && r2 != 0 && r2 % 2 == 0);
-   extension(S390_EXT_KMC, r1 | (r2 << 4));
-   return "kmc";
-}
-
-static const HChar *
-s390_irgen_KIMD(UChar r1, UChar r2)
-{
-   /* r1 is reserved */
-   s390_insn_assert("kimd", r2 != 0 && r2 % 2 == 0);
-   extension(S390_EXT_KIMD, r1 | (r2 << 4));
-   return "kimd";
-}
-
-static const HChar *
-s390_irgen_KLMD(UChar r1, UChar r2)
-{
-   /* r1 is only used by some functions */
-   s390_insn_assert("klmd", r2 != 0 && r2 % 2 == 0);
-   extension(S390_EXT_KLMD, r1 | (r2 << 4));
-   return "klmd";
-}
-
-static const HChar *
-s390_irgen_KMAC(UChar r1, UChar r2)
-{
-   /* r1 is ignored */
-   s390_insn_assert("kmac", r2 != 0 && r2 % 2 == 0);
-   extension(S390_EXT_KMAC, r1 | (r2 << 4));
-   return "kmac";
-}
-
-static const HChar *
-s390_irgen_PCC(void)
-{
-   extension(S390_EXT_PCC, 0);
-   return "pcc";
-}
-
-static const HChar *
-s390_irgen_KMCTR(UChar r3, UChar r1, UChar r2)
-{
-   s390_insn_assert("kmctr", r1 % 2 == 0 && r1 != 0 && r2 % 2 == 0 && r2 != 0 &&
-                    r3 % 2 == 0 && r3 != 0);
-   extension(S390_EXT_KMCTR, r1 | (r2 << 4) | (r3 << 8));
-   return "kmctr";
-}
-
-static const HChar *
-s390_irgen_KMO(UChar r1, UChar r2)
-{
-   s390_insn_assert("kmo", r1 != 0 && r1 % 2 == 0 && r2 != 0 && r2 % 2 == 0);
-   extension(S390_EXT_KMO, r1 | (r2 << 4));
-   return "kmo";
-}
-
-static const HChar *
-s390_irgen_KMF(UChar r1, UChar r2)
-{
-   s390_insn_assert("kmf", r1 != 0 && r1 % 2 == 0 && r2 != 0 && r2 % 2 == 0);
-   extension(S390_EXT_KMF, r1 | (r2 << 4));
-   return "kmf";
-}
-
-static const HChar *
-s390_irgen_KMA(UChar r3, UChar r1, UChar r2)
-{
-   s390_insn_assert("kma", r1 % 2 == 0 && r1 != 0 && r2 % 2 == 0 && r2 != 0 &&
-                    r3 % 2 == 0 && r3 != 0);
-   extension(S390_EXT_KMA, r1 | (r2 << 4) | (r3 << 8));
-   return "kma";
-}
-
-static const HChar *
-s390_irgen_KDSA(UChar r1, UChar r2)
-{
-   /* r1 is reserved */
-   s390_insn_assert("kdsa", r2 != 0 && r2 % 2 == 0);
-   extension(S390_EXT_KDSA, r1 | (r2 << 4));
-   return "kdsa";
-}
-
 /* New insns are added here.
    If an insn is contingent on a facility being installed also
    check whether the list of supported facilities in function
@@ -19909,10 +19803,10 @@ s390_decode_2byte_and_irgen(const UChar *bytes)
    case 0x0c: /* BASSM */ goto unimplemented;
    case 0x0d: s390_format_RR_RR(s390_irgen_BASR, RR_r1(ovl), RR_r2(ovl));
                                 goto ok;
-   case 0x0e: s390_format_RR_RR(s390_irgen_MVCL, RR_r1(ovl), RR_r2(ovl));
-                                goto ok;
-   case 0x0f: s390_format_RR_RR(s390_irgen_CLCL, RR_r1(ovl), RR_r2(ovl));
-                                goto ok;
+   case 0x0e: s390_format_RR(s390_irgen_MVCL, RR_r1(ovl), RR_r2(ovl));
+                             goto ok;
+   case 0x0f: s390_format_RR(s390_irgen_CLCL, RR_r1(ovl), RR_r2(ovl));
+                             goto ok;
    case 0x10: s390_format_RR_RR(s390_irgen_LPR, RR_r1(ovl), RR_r2(ovl));
                                 goto ok;
    case 0x11: s390_format_RR_RR(s390_irgen_LNR, RR_r1(ovl), RR_r2(ovl));
@@ -20067,8 +19961,6 @@ s390_decode_4byte_and_irgen(const UChar *bytes)
    case 0x8000: /* SSM */ goto unimplemented;
    case 0x8200: /* LPSW */ goto unimplemented;
    case 0x9300: /* TS */ goto unimplemented;
-   case 0xb200: /* LBEAR */ goto unimplemented;
-   case 0xb201: /* STBEAR */ goto unimplemented;
    case 0xb202: /* STIDP */ goto unimplemented;
    case 0xb204: /* SCK */ goto unimplemented;
    case 0xb205: s390_format_S_RD(s390_irgen_STCK, S_b2(ovl), S_d2(ovl));
@@ -20116,8 +20008,8 @@ s390_decode_4byte_and_irgen(const UChar *bytes)
    case 0xb23b: /* RCHP */ goto unimplemented;
    case 0xb23c: /* SCHM */ goto unimplemented;
    case 0xb240: /* BAKR */ goto unimplemented;
-   case 0xb241: s390_format_RRE_RR(s390_irgen_CKSM, RRE_r1(ovl),
-                                   RRE_r2(ovl));  goto ok;
+   case 0xb241: s390_format_RRE(s390_irgen_CKSM, RRE_r1(ovl),
+                                RRE_r2(ovl));  goto ok;
    case 0xb244: /* SQDR */ goto unimplemented;
    case 0xb245: /* SQER */ goto unimplemented;
    case 0xb246: /* STURA */ goto unimplemented;
@@ -20160,7 +20052,6 @@ s390_decode_4byte_and_irgen(const UChar *bytes)
    case 0xb286: /* QSI */ goto unimplemented;
    case 0xb287: /* LSCTL */ goto unimplemented;
    case 0xb28e: /* QCTRI */ goto unimplemented;
-   case 0xb28f: /* QPACI */ goto unimplemented;
    case 0xb299: s390_format_S_RD(s390_irgen_SRNM, S_b2(ovl), S_d2(ovl));
                                  goto ok;
    case 0xb29c: s390_format_S_RD(s390_irgen_STFPC, S_b2(ovl), S_d2(ovl));
@@ -20174,7 +20065,7 @@ s390_decode_4byte_and_irgen(const UChar *bytes)
    case 0xb2a7: s390_format_RRF_M0RERE(s390_irgen_CU12, RRF3_r3(ovl),
                                        RRF3_r1(ovl), RRF3_r2(ovl));
       goto ok;
-   case 0xb2b0: s390_format_S_RD_raw(s390_irgen_STFLE, S_b2(ovl), S_d2(ovl));
+   case 0xb2b0: s390_format_S_RD(s390_irgen_STFLE, S_b2(ovl), S_d2(ovl));
                                  goto ok;
    case 0xb2b1: /* STFL */ goto unimplemented;
    case 0xb2b2: /* LPSWE */ goto unimplemented;
@@ -20580,8 +20471,7 @@ s390_decode_4byte_and_irgen(const UChar *bytes)
                                    RRE_r2(ovl));  goto ok;
    case 0xb91d: s390_format_RRE_RR(s390_irgen_DSGFR, RRE_r1(ovl),
                                    RRE_r2(ovl));  goto ok;
-   case 0xb91e: s390_format_RRE_RR(s390_irgen_KMAC, RRE_r1(ovl),
-                                   RRE_r2(ovl));  goto ok;
+   case 0xb91e: /* KMAC */ goto unimplemented;
    case 0xb91f: s390_format_RRE_RR(s390_irgen_LRVR, RRE_r1(ovl),
                                    RRE_r2(ovl));  goto ok;
    case 0xb920: s390_format_RRE_RR(s390_irgen_CGR, RRE_r1(ovl),
@@ -20594,35 +20484,24 @@ s390_decode_4byte_and_irgen(const UChar *bytes)
    case 0xb927: s390_format_RRE_RR(s390_irgen_LHR, RRE_r1(ovl),
                                    RRE_r2(ovl));  goto ok;
    case 0xb928: /* PCKMO */ goto unimplemented;
-   case 0xb929: s390_format_RRF_R0RR2(s390_irgen_KMA, RRF4_r3(ovl),
-                                      RRF4_r1(ovl), RRF4_r2(ovl)); goto ok;
-   case 0xb92a: s390_format_RRE_RR(s390_irgen_KMF, RRE_r1(ovl),
-                                   RRE_r2(ovl));  goto ok;
-   case 0xb92b: s390_format_RRE_RR(s390_irgen_KMO, RRE_r1(ovl),
-                                   RRE_r2(ovl));  goto ok;
-   case 0xb92c: s390_format_E(s390_irgen_PCC);  goto ok;
-   case 0xb92d: s390_format_RRF_R0RR2(s390_irgen_KMCTR, RRF4_r3(ovl),
-                                      RRF4_r1(ovl), RRF4_r2(ovl)); goto ok;
-   case 0xb92e: s390_format_RRE_RR(s390_irgen_KM, RRE_r1(ovl),
-                                   RRE_r2(ovl));  goto ok;
-   case 0xb92f: s390_format_RRE_RR(s390_irgen_KMC, RRE_r1(ovl),
-                                   RRE_r2(ovl));  goto ok;
+   case 0xb929: /* KMA */ goto unimplemented;
+   case 0xb92a: /* KMF */ goto unimplemented;
+   case 0xb92b: /* KMO */ goto unimplemented;
+   case 0xb92c: /* PCC */ goto unimplemented;
+   case 0xb92d: /* KMCTR */ goto unimplemented;
+   case 0xb92e: /* KM */ goto unimplemented;
+   case 0xb92f: /* KMC */ goto unimplemented;
    case 0xb930: s390_format_RRE_RR(s390_irgen_CGFR, RRE_r1(ovl),
                                    RRE_r2(ovl));  goto ok;
    case 0xb931: s390_format_RRE_RR(s390_irgen_CLGFR, RRE_r1(ovl),
                                    RRE_r2(ovl));  goto ok;
    case 0xb938: /* SORTL */ goto unimplemented;
-   case 0xb939: s390_format_RRF_R0RR2(s390_irgen_DFLTCC, RRF4_r3(ovl),
-                                      RRF4_r1(ovl), RRF4_r2(ovl)); goto ok;
-   case 0xb93a: s390_format_RRE_RR(s390_irgen_KDSA, RRE_r1(ovl),
-                                   RRE_r2(ovl));  goto ok;
-   case 0xb93b: s390_format_E(s390_irgen_NNPA);  goto ok;
+   case 0xb939: /* DFLTCC */ goto unimplemented;
+   case 0xb93a: /* KDSA */ goto unimplemented;
    case 0xb93c: s390_format_RRE_RR(s390_irgen_PPNO, RRE_r1(ovl),
                                    RRE_r2(ovl));  goto ok;
-   case 0xb93e: s390_format_RRE_RR(s390_irgen_KIMD, RRE_r1(ovl),
-                                   RRE_r2(ovl));  goto ok;
-   case 0xb93f: s390_format_RRE_RR(s390_irgen_KLMD, RRE_r1(ovl),
-                                   RRE_r2(ovl));  goto ok;
+   case 0xb93e: /* KIMD */ goto unimplemented;
+   case 0xb93f: /* KLMD */ goto unimplemented;
    case 0xb941: s390_format_RRF_UURF(s390_irgen_CFDTR, RRF2_m3(ovl),
                                      RRF2_m4(ovl), RRF2_r1(ovl),
                                      RRF2_r2(ovl));  goto ok;
@@ -20675,10 +20554,6 @@ s390_decode_4byte_and_irgen(const UChar *bytes)
                                       RRF4_r1(ovl), RRF4_r2(ovl)); goto ok;
    case 0xb967: s390_format_RRF_R0RR2(s390_irgen_NXGRK, RRF4_r3(ovl),
                                       RRF4_r1(ovl), RRF4_r2(ovl)); goto ok;
-   case 0xb968: /* CLZG */ goto unimplemented;
-   case 0xb969: /* CTZG */ goto unimplemented;
-   case 0xb96c: /* BEXTG */ goto unimplemented;
-   case 0xb96d: /* BDEPG */ goto unimplemented;
    case 0xb972: s390_format_RRF_U0RR(s390_irgen_CRT, RRF2_m3(ovl),
                                      RRF2_r1(ovl), RRF2_r2(ovl),
                                      S390_XMNM_CAB); goto ok;
@@ -20714,7 +20589,6 @@ s390_decode_4byte_and_irgen(const UChar *bytes)
    case 0xb989: s390_format_RRE_RR(s390_irgen_SLBGR, RRE_r1(ovl),
                                    RRE_r2(ovl));  goto ok;
    case 0xb98a: /* CSPG */ goto unimplemented;
-   case 0xb98b: /* RDP */ goto unimplemented;
    case 0xb98d: /* EPSW */ goto unimplemented;
    case 0xb98e: /* IDTE */ goto unimplemented;
    case 0xb98f: /* CRDTE */ goto unimplemented;
@@ -20799,12 +20673,12 @@ s390_decode_4byte_and_irgen(const UChar *bytes)
                                    RRE_r2(ovl));  goto ok;
    case 0xb9e0: s390_format_RRF_U0RR(s390_irgen_LOCFHR, RRF3_r3(ovl),
                                      RRF3_r1(ovl), RRF3_r2(ovl),
-                                     S390_XMNM_CLS);  goto ok;
+                                     S390_XMNM_LOCFHR);  goto ok;
    case 0xb9e1: s390_format_RRFa_U0RR(s390_irgen_POPCNT, RRF3_r3(ovl),
                                       RRF3_r1(ovl), RRF3_r2(ovl));  goto ok;
    case 0xb9e2: s390_format_RRF_U0RR(s390_irgen_LOCGR, RRF3_r3(ovl),
                                      RRF3_r1(ovl), RRF3_r2(ovl),
-                                     S390_XMNM_CLS);  goto ok;
+                                     S390_XMNM_LOCGR);  goto ok;
    case 0xb9e3: s390_format_RRF_RURR(s390_irgen_SELGR, RRF4_r3(ovl),
                                      RRF4_m4(ovl), RRF4_r1(ovl),
                                      RRF4_r2(ovl)); goto ok;
@@ -20843,7 +20717,7 @@ s390_decode_4byte_and_irgen(const UChar *bytes)
                                      RRF4_r2(ovl)); goto ok;
    case 0xb9f2: s390_format_RRF_U0RR(s390_irgen_LOCR, RRF3_r3(ovl),
                                      RRF3_r1(ovl), RRF3_r2(ovl),
-                                     S390_XMNM_CLS);  goto ok;
+                                     S390_XMNM_LOCR);  goto ok;
    case 0xb9f4: s390_format_RRF_R0RR2(s390_irgen_NRK, RRF4_r3(ovl),
                                       RRF4_r1(ovl), RRF4_r2(ovl));
                                       goto ok;
@@ -21283,16 +21157,6 @@ s390_decode_6byte_and_irgen(const UChar *bytes)
                                                 RXY_x2(ovl), RXY_b2(ovl),
                                                 RXY_dl2(ovl),
                                                 RXY_dh2(ovl));  goto ok;
-   case 0xe30000000060ULL: /* LXAB */ goto unimplemented;
-   case 0xe30000000061ULL: /* LLXAB */ goto unimplemented;
-   case 0xe30000000062ULL: /* LXAH */ goto unimplemented;
-   case 0xe30000000063ULL: /* LLXAH */ goto unimplemented;
-   case 0xe30000000064ULL: /* LXAF */ goto unimplemented;
-   case 0xe30000000065ULL: /* LLXAF */ goto unimplemented;
-   case 0xe30000000066ULL: /* LXAG */ goto unimplemented;
-   case 0xe30000000067ULL: /* LLXAG */ goto unimplemented;
-   case 0xe30000000068ULL: /* LXAQ */ goto unimplemented;
-   case 0xe30000000069ULL: /* LLXAQ */ goto unimplemented;
    case 0xe30000000070ULL: s390_format_RXY_RRRD(s390_irgen_STHY, RXY_r1(ovl),
                                                 RXY_x2(ovl), RXY_b2(ovl),
                                                 RXY_dl2(ovl),
@@ -21550,53 +21414,21 @@ s390_decode_6byte_and_irgen(const UChar *bytes)
                                                 VRS_d2(ovl),
                                                 VRS_rxb(ovl));  goto ok;
    case 0xe60000000049ULL: /* VLIP */ goto unimplemented;
-   case 0xe6000000004aULL: /* VCVDQ */ goto unimplemented;
-   case 0xe6000000004eULL: /* VCVBQ */ goto unimplemented;
    case 0xe60000000050ULL: /* VCVB */ goto unimplemented;
-   case 0xe60000000051ULL: /* VCLZDP */ goto unimplemented;
    case 0xe60000000052ULL: /* VCVBG */ goto unimplemented;
-   case 0xe60000000054ULL: /* VUPKZH */ goto unimplemented;
-   case 0xe60000000055ULL: s390_format_VRRa_VVMM(s390_irgen_VCNF,
-                                                 VRRa_v1(ovl), VRRa_v2(ovl),
-                                                 VRRa_m3(ovl), VRRa_m4(ovl),
-                                                 VRRa_rxb(ovl));  goto ok;
-   case 0xe60000000056ULL: s390_format_VRRa_VVMM(s390_irgen_VCLFNH,
-                                                 VRRa_v1(ovl), VRRa_v2(ovl),
-                                                 VRRa_m3(ovl), VRRa_m4(ovl),
-                                                 VRRa_rxb(ovl));  goto ok;
-   case 0xe6000000005dULL: s390_format_VRRa_VVMM(s390_irgen_VCFN,
-                                                 VRRa_v1(ovl), VRRa_v2(ovl),
-                                                 VRRa_m3(ovl), VRRa_m4(ovl),
-                                                 VRRa_rxb(ovl));  goto ok;
-   case 0xe6000000005eULL: s390_format_VRRa_VVMM(s390_irgen_VCLFNL,
-                                                 VRRa_v1(ovl), VRRa_v2(ovl),
-                                                 VRRa_m3(ovl), VRRa_m4(ovl),
-                                                 VRRa_rxb(ovl));  goto ok;
    case 0xe60000000058ULL: /* VCVD */ goto unimplemented;
    case 0xe60000000059ULL: /* VSRP */ goto unimplemented;
    case 0xe6000000005aULL: /* VCVDG */ goto unimplemented;
    case 0xe6000000005bULL: /* VPSOP */ goto unimplemented;
-   case 0xe6000000005cULL: /* VUPKZL */ goto unimplemented;
    case 0xe6000000005fULL: /* VTP */ goto unimplemented;
-   case 0xe60000000070ULL: /* VPKZR */ goto unimplemented;
    case 0xe60000000071ULL: /* VAP */ goto unimplemented;
-   case 0xe60000000072ULL: /* VSRPR */ goto unimplemented;
    case 0xe60000000073ULL: /* VSP */ goto unimplemented;
-   case 0xe60000000075ULL: s390_format_VRRa_VVVMM(s390_irgen_VCRNF,
-                                                  VRRa_v1(ovl), VRRa_v2(ovl),
-                                                  VRRa_v3(ovl),
-                                                  VRRa_m3(ovl), VRRa_m4(ovl),
-                                                  VRRa_rxb(ovl)); goto ok;
    case 0xe60000000077ULL: /* VCP */ goto unimplemented;
    case 0xe60000000078ULL: /* VMP */ goto unimplemented;
    case 0xe60000000079ULL: /* VMSP */ goto unimplemented;
-   case 0xe60000000074ULL: /* VSCHP */ goto unimplemented;
    case 0xe6000000007aULL: /* VDP */ goto unimplemented;
    case 0xe6000000007bULL: /* VRP */ goto unimplemented;
-   case 0xe6000000007cULL: /* VSCSHP */ goto unimplemented;
-   case 0xe6000000007dULL: /* VCSPH */ goto unimplemented;
    case 0xe6000000007eULL: /* VSDP */ goto unimplemented;
-   case 0xe6000000007fULL: /* VTZ */ goto unimplemented;
    case 0xe70000000000ULL: s390_format_VRX_VRRDM(s390_irgen_VLEB, VRX_v1(ovl),
                                                  VRX_x2(ovl), VRX_b2(ovl),
                                                  VRX_d2(ovl), VRX_m3(ovl),
@@ -21737,7 +21569,6 @@ s390_decode_6byte_and_irgen(const UChar *bytes)
    case 0xe70000000053ULL: s390_format_VRR_VVM(s390_irgen_VCLZ, VRR_v1(ovl),
                                                VRR_v2(ovl), VRR_m4(ovl),
                                                VRR_rxb(ovl));  goto ok;
-   case 0xe70000000054ULL: /* VGEM */ goto unimplemented;
    case 0xe70000000056ULL: s390_format_VRR_VV(s390_irgen_VLR, VRR_v1(ovl),
                                               VRR_v2(ovl), VRR_rxb(ovl));  goto ok;
    case 0xe7000000005cULL: s390_format_VRR_VVMM(s390_irgen_VISTR, VRR_v1(ovl),
@@ -21854,8 +21685,6 @@ s390_decode_6byte_and_irgen(const UChar *bytes)
                                                  VRId_v2(ovl), VRId_v3(ovl),
                                                  VRId_i4(ovl),
                                                  VRId_rxb(ovl));  goto ok;
-   case 0xe70000000088ULL: /* VEVAL */ goto unimplemented;
-   case 0xe70000000089ULL: /* VBLEND */ goto unimplemented;
    case 0xe7000000008aULL: s390_format_VRR_VVVVMM(s390_irgen_VSTRC, VRRd_v1(ovl),
                                                   VRRd_v2(ovl), VRRd_v3(ovl),
                                                   VRRd_v4(ovl), VRRd_m5(ovl),
@@ -21950,10 +21779,6 @@ s390_decode_6byte_and_irgen(const UChar *bytes)
                                                   VRRd_v2(ovl), VRRd_v3(ovl),
                                                   VRRd_v4(ovl), VRRd_m5(ovl),
                                                   VRRd_rxb(ovl));  goto ok;
-   case 0xe700000000b0ULL: /* VDL */ goto unimplemented;
-   case 0xe700000000b1ULL: /* VRL */ goto unimplemented;
-   case 0xe700000000b2ULL: /* VD */ goto unimplemented;
-   case 0xe700000000b3ULL: /* VR */ goto unimplemented;
    case 0xe700000000b4ULL: s390_format_VRR_VVVM(s390_irgen_VGFM, VRR_v1(ovl),
                                                 VRR_v2(ovl), VRR_r3(ovl),
                                                 VRR_m4(ovl), VRR_rxb(ovl));  goto ok;
@@ -22160,7 +21985,6 @@ s390_decode_6byte_and_irgen(const UChar *bytes)
                                                 RSY_r3(ovl), RSY_b2(ovl),
                                                 RSY_dl2(ovl),
                                                 RSY_dh2(ovl));  goto ok;
-   case 0xeb0000000016ULL: /* PFCR */ goto unimplemented;
    case 0xeb000000001cULL: s390_format_RSY_RRRD(s390_irgen_RLLG, RSY_r1(ovl),
                                                 RSY_r3(ovl), RSY_b2(ovl),
                                                 RSY_dl2(ovl),
@@ -22177,7 +22001,7 @@ s390_decode_6byte_and_irgen(const UChar *bytes)
                                                 RSY_r3(ovl), RSY_b2(ovl),
                                                 RSY_dl2(ovl),
                                                 RSY_dh2(ovl));  goto ok;
-   case 0xeb0000000023ULL: s390_format_RSY_R0RD(s390_irgen_CLT, RSY_r1(ovl),
+   case 0xeb0000000023ULL: s390_format_RSY_RURD(s390_irgen_CLT, RSY_r1(ovl),
                                                 RSY_r3(ovl), RSY_b2(ovl),
                                                 RSY_dl2(ovl),
                                                 RSY_dh2(ovl));  goto ok;
@@ -22190,7 +22014,7 @@ s390_decode_6byte_and_irgen(const UChar *bytes)
                                                 RSY_r3(ovl), RSY_b2(ovl),
                                                 RSY_dl2(ovl),
                                                 RSY_dh2(ovl));  goto ok;
-   case 0xeb000000002bULL: s390_format_RSY_R0RD(s390_irgen_CLGT, RSY_r1(ovl),
+   case 0xeb000000002bULL: s390_format_RSY_RURD(s390_irgen_CLGT, RSY_r1(ovl),
                                                 RSY_r3(ovl), RSY_b2(ovl),
                                                 RSY_dl2(ovl),
                                                 RSY_dh2(ovl));  goto ok;
@@ -22251,7 +22075,6 @@ s390_decode_6byte_and_irgen(const UChar *bytes)
    case 0xeb000000006eULL: s390_format_SIY_IRD(s390_irgen_ALSI, SIY_i2(ovl),
                                                SIY_b1(ovl), SIY_dl1(ovl),
                                                SIY_dh1(ovl));  goto ok;
-   case 0xeb0000000071ULL: /* LPSWEY */ goto unimplemented;
    case 0xeb000000007aULL: s390_format_SIY_IRD(s390_irgen_AGSI, SIY_i2(ovl),
                                                SIY_b1(ovl), SIY_dl1(ovl),
                                                SIY_dh1(ovl));  goto ok;
@@ -22308,19 +22131,23 @@ s390_decode_6byte_and_irgen(const UChar *bytes)
    case 0xeb00000000e0ULL: s390_format_RSY_RDRM(s390_irgen_LOCFH, RSY_r1(ovl),
                                                 RSY_r3(ovl), RSY_b2(ovl),
                                                 RSY_dl2(ovl),
-                                                RSY_dh2(ovl));  goto ok;
+                                                RSY_dh2(ovl),
+                                                S390_XMNM_LOCFH);  goto ok;
    case 0xeb00000000e1ULL: s390_format_RSY_RDRM(s390_irgen_STOCFH, RSY_r1(ovl),
                                                 RSY_r3(ovl), RSY_b2(ovl),
                                                 RSY_dl2(ovl),
-                                                RSY_dh2(ovl));  goto ok;
+                                                RSY_dh2(ovl),
+                                                S390_XMNM_STOCFH);  goto ok;
    case 0xeb00000000e2ULL: s390_format_RSY_RDRM(s390_irgen_LOCG, RSY_r1(ovl),
                                                 RSY_r3(ovl), RSY_b2(ovl),
                                                 RSY_dl2(ovl),
-                                                RSY_dh2(ovl));  goto ok;
+                                                RSY_dh2(ovl),
+                                                S390_XMNM_LOCG);  goto ok;
    case 0xeb00000000e3ULL: s390_format_RSY_RDRM(s390_irgen_STOCG,
                                                 RSY_r1(ovl), RSY_r3(ovl),
                                                 RSY_b2(ovl), RSY_dl2(ovl),
-                                                RSY_dh2(ovl));  goto ok;
+                                                RSY_dh2(ovl),
+                                                S390_XMNM_STOCG);  goto ok;
    case 0xeb00000000e4ULL: s390_format_RSY_RRRD(s390_irgen_LANG, RSY_r1(ovl),
                                                 RSY_r3(ovl), RSY_b2(ovl),
                                                 RSY_dl2(ovl),
@@ -22344,11 +22171,13 @@ s390_decode_6byte_and_irgen(const UChar *bytes)
    case 0xeb00000000f2ULL: s390_format_RSY_RDRM(s390_irgen_LOC, RSY_r1(ovl),
                                                 RSY_r3(ovl), RSY_b2(ovl),
                                                 RSY_dl2(ovl),
-                                                RSY_dh2(ovl)); goto ok;
+                                                RSY_dh2(ovl), S390_XMNM_LOC);
+                                                goto ok;
    case 0xeb00000000f3ULL: s390_format_RSY_RDRM(s390_irgen_STOC, RSY_r1(ovl),
                                                 RSY_r3(ovl), RSY_b2(ovl),
                                                 RSY_dl2(ovl),
-                                                RSY_dh2(ovl));  goto ok;
+                                                RSY_dh2(ovl),
+                                                S390_XMNM_STOC);  goto ok;
    case 0xeb00000000f4ULL: s390_format_RSY_RRRD(s390_irgen_LAN, RSY_r1(ovl),
                                                 RSY_r3(ovl), RSY_b2(ovl),
                                                 RSY_dl2(ovl),
@@ -22372,7 +22201,9 @@ s390_decode_6byte_and_irgen(const UChar *bytes)
    case 0xec0000000042ULL: s390_format_RIE_RUPIX(s390_irgen_LOCHI,
                                                  RIEv3_r1(ovl),
                                                  RIEv3_m3(ovl),
-                                                 RIEv3_i4(ovl));  goto ok;
+                                                 RIEv3_i4(ovl),
+                                                 RIEv3_i2(ovl),
+                                                 S390_XMNM_LOCHI);  goto ok;
    case 0xec0000000044ULL: s390_format_RIE_RRP(s390_irgen_BRXHG, RIE_r1(ovl),
                                                RIE_r3(ovl), RIE_i2(ovl));
                                                goto ok;
@@ -22382,11 +22213,15 @@ s390_decode_6byte_and_irgen(const UChar *bytes)
    case 0xec0000000046ULL: s390_format_RIE_RUPIX(s390_irgen_LOCGHI,
                                                  RIEv3_r1(ovl),
                                                  RIEv3_m3(ovl),
-                                                 RIEv3_i4(ovl));  goto ok;
+                                                 RIEv3_i4(ovl),
+                                                 RIEv3_i2(ovl),
+                                                 S390_XMNM_LOCGHI);  goto ok;
    case 0xec000000004eULL: s390_format_RIE_RUPIX(s390_irgen_LOCHHI,
                                                  RIEv3_r1(ovl),
                                                  RIEv3_m3(ovl),
-                                                 RIEv3_i4(ovl));  goto ok;
+                                                 RIEv3_i4(ovl),
+                                                 RIEv3_i2(ovl),
+                                                 S390_XMNM_LOCHHI);  goto ok;
    case 0xec0000000051ULL: s390_format_RIE_RRUUU(s390_irgen_RISBLG,
                                                  RIE_RRUUU_r1(ovl),
                                                  RIE_RRUUU_r2(ovl),
@@ -22446,22 +22281,22 @@ s390_decode_6byte_and_irgen(const UChar *bytes)
                                                 RIE_RRPU_r2(ovl),
                                                 RIE_RRPU_i4(ovl),
                                                 RIE_RRPU_m3(ovl));  goto ok;
-   case 0xec0000000070ULL: s390_format_R0IU(s390_irgen_CGIT,
-                                            RIE_R0xU_r1(ovl),
-                                            RIE_R0xU_i2(ovl),
-                                            RIE_R0xU_m3(ovl)); goto ok;
-   case 0xec0000000071ULL: s390_format_R0UU(s390_irgen_CLGIT,
-                                            RIE_R0xU_r1(ovl),
-                                            RIE_R0xU_i2(ovl),
-                                            RIE_R0xU_m3(ovl)); goto ok;
-   case 0xec0000000072ULL: s390_format_R0IU(s390_irgen_CIT,
-                                            RIE_R0xU_r1(ovl),
-                                            RIE_R0xU_i2(ovl),
-                                            RIE_R0xU_m3(ovl)); goto ok;
-   case 0xec0000000073ULL: s390_format_R0UU(s390_irgen_CLFIT,
-                                            RIE_R0xU_r1(ovl),
-                                            RIE_R0xU_i2(ovl),
-                                            RIE_R0xU_m3(ovl)); goto ok;
+   case 0xec0000000070ULL: s390_format_RIEv1(s390_irgen_CGIT,
+                                             RIEv1_r1(ovl),
+                                             RIEv1_i2(ovl),
+                                             RIEv1_m3(ovl)); goto ok;
+   case 0xec0000000071ULL: s390_format_RIEv1(s390_irgen_CLGIT,
+                                             RIEv1_r1(ovl),
+                                             RIEv1_i2(ovl),
+                                             RIEv1_m3(ovl)); goto ok;
+   case 0xec0000000072ULL: s390_format_RIEv1(s390_irgen_CIT,
+                                             RIEv1_r1(ovl),
+                                             RIEv1_i2(ovl),
+                                             RIEv1_m3(ovl)); goto ok;
+   case 0xec0000000073ULL: s390_format_RIEv1(s390_irgen_CLFIT,
+                                             RIEv1_r1(ovl),
+                                             RIEv1_i2(ovl),
+                                             RIEv1_m3(ovl)); goto ok;
    case 0xec0000000076ULL: s390_format_RIE_RRPU(s390_irgen_CRJ,
                                                 RIE_RRPU_r1(ovl),
                                                 RIE_RRPU_r2(ovl),
@@ -22797,9 +22632,6 @@ s390_decode_6byte_and_irgen(const UChar *bytes)
    case 0xc802ULL: /* CSST */ goto unimplemented;
    case 0xc804ULL: /* LPD */ goto unimplemented;
    case 0xc805ULL: /* LPDG */ goto unimplemented;
-   case 0xc806ULL: /* CAL */ goto unimplemented;
-   case 0xc807ULL: /* CALG */ goto unimplemented;
-   case 0xc80fULL: /* CALGF */ goto unimplemented;
    case 0xcc06ULL:  s390_format_RIL_RP(s390_irgen_BRCTH, RIL_r1(ovl),
                                        RIL_i2(ovl));  goto ok;
    case 0xcc08ULL: s390_format_RIL_RI(s390_irgen_AIH, RIL_r1(ovl),
@@ -23005,9 +22837,7 @@ s390_decode_and_irgen(const UChar *bytes, UInt insn_length, DisResult *dres)
       }
    }
    /* If next instruction is execute, stop here */
-   if (dis_res->whatNext == Dis_Continue &&
-       (bytes[insn_length] == 0x44 ||
-        (bytes[insn_length] == 0xc6 && (bytes[insn_length + 1] & 0xf) == 0))) {
+   if (dis_res->whatNext == Dis_Continue && bytes[insn_length] == 0x44) {
       put_IA(mkaddr_expr(guest_IA_next_instr));
       dis_res->whatNext = Dis_StopHere;
       dis_res->jk_StopHere = Ijk_Boring;
@@ -23154,8 +22984,6 @@ disInstr_S390(IRSB        *irsb_IN,
 
    /* Set globals (see top of this file) */
    guest_IA_curr_instr = guest_IP;
-   if (last_execute_target == Invalid_execute_target)
-      guest_IA_rel_base = guest_IA_curr_instr;
    irsb = irsb_IN;
    sigill_diag = sigill_diag_IN;
 
