diff --git a/coregrind/m_libcfile.c b/coregrind/m_libcfile.c
index fb93b4867..40a1e0e67 100644
--- a/coregrind/m_libcfile.c
+++ b/coregrind/m_libcfile.c
@@ -93,7 +93,7 @@ static Char filedesc_buf[M_FILEDESC_BUF];
    cannot be deduced. */
 Bool VG_(resolve_filename) ( Int fd, const HChar** result )
 {
-#  if defined(VGO_linux) || defined(VGO_solaris)
+#  if defined(VGO_linux) || defined(VGO_solaris) || defined(VGO_openedition)
    static HChar *buf = NULL;
    static SizeT  bufsiz = 0;
 
@@ -104,7 +104,7 @@ Bool VG_(resolve_filename) ( Int fd, const HChar** result )
 
    HChar tmp[64];   // large enough
    {
-#     if defined(VGO_linux)
+#     if defined(VGO_linux) || defined(VGO_openedition)
       VG_(sprintf)(tmp, "/proc/self/fd/%d", fd);
 #     elif defined(VGO_solaris)
       VG_(sprintf)(tmp, "/proc/self/path/%d", fd);
@@ -268,7 +268,7 @@ SysRes VG_(mknod) ( const HChar* pathname, Int mode, UWord dev )
    /* ARM64 wants to use __NR_mknodat rather than __NR_mknod. */
    SysRes res = VG_(do_syscall4)(__NR_mknodat,
                                  VKI_AT_FDCWD, (UWord)pathname, mode, dev);
-#  elif defined(VGO_linux) || defined(VGO_darwin)
+#  elif defined(VGO_linux) || defined(VGO_darwin) || defined(VGO_openedition)
    SysRes res = VG_(do_syscall3)(__NR_mknod,
                                  (UWord)pathname, mode, dev);
 #  elif defined(VGO_freebsd)
@@ -294,7 +294,7 @@ SysRes VG_(open) ( const HChar* pathname, Int flags, Int mode )
    /* ARM64 wants to use __NR_openat rather than __NR_open. */
    SysRes res = VG_(do_syscall4)(__NR_openat,
                                  VKI_AT_FDCWD, (UWord)pathname, flags, mode);
-#  elif defined(VGO_linux) || defined(VGO_freebsd)
+#  elif defined(VGO_linux) || defined(VGO_freebsd) || defined(VGO_openedition)
    SysRes res = VG_(do_syscall3)(__NR_open,
                                  (UWord)pathname, flags, mode);
 #  elif defined(VGO_darwin)
@@ -322,7 +322,8 @@ Int VG_(fd_open) (const HChar* pathname, Int flags, Int mode)
 void VG_(close) ( Int fd )
 {
    /* Hmm.  Return value is not checked.  That's uncool. */
-#  if defined(VGO_linux) || defined(VGO_solaris) || defined(VGO_freebsd)
+#  if defined(VGO_linux) || defined(VGO_solaris) || defined(VGO_freebsd) \
+      || defined(VGO_openedition)
    (void)VG_(do_syscall1)(__NR_close, fd);
 #  elif defined(VGO_darwin)
    (void)VG_(do_syscall1)(__NR_close_nocancel, fd);
@@ -334,7 +335,8 @@ void VG_(close) ( Int fd )
 Int VG_(read) ( Int fd, void* buf, Int count)
 {
    Int    ret;
-#  if defined(VGO_linux) || defined(VGO_solaris) || defined(VGO_freebsd)
+#  if defined(VGO_linux) || defined(VGO_solaris) || defined(VGO_freebsd) \
+      || defined(VGO_openedition)
    SysRes res = VG_(do_syscall3)(__NR_read, fd, (UWord)buf, count);
 #  elif defined(VGO_darwin)
    SysRes res = VG_(do_syscall3)(__NR_read_nocancel, fd, (UWord)buf, count);
@@ -354,7 +356,8 @@ Int VG_(read) ( Int fd, void* buf, Int count)
 Int VG_(write) ( Int fd, const void* buf, Int count)
 {
    Int    ret;
-#  if defined(VGO_linux) || defined(VGO_solaris) || defined(VGO_freebsd)
+#  if defined(VGO_linux) || defined(VGO_solaris) || defined(VGO_freebsd) \
+      || defined(VGO_openedition)
    SysRes res = VG_(do_syscall3)(__NR_write, fd, (UWord)buf, count);
 #  elif defined(VGO_darwin)
    SysRes res = VG_(do_syscall3)(__NR_write_nocancel, fd, (UWord)buf, count);
@@ -387,7 +390,7 @@ Int VG_(pipe) ( Int fd[2] )
 #  elif defined(VGP_arm64_linux) || defined(VGP_nanomips_linux)
    SysRes res = VG_(do_syscall2)(__NR_pipe2, (UWord)fd, 0);
    return sr_isError(res) ? -1 : 0;
-#  elif defined(VGO_linux)
+#  elif defined(VGO_linux) || defined(VGO_openedition)
    SysRes res = VG_(do_syscall1)(__NR_pipe, (UWord)fd);
    return sr_isError(res) ? -1 : 0;
 #  elif defined(VGO_freebsd)
@@ -428,7 +431,8 @@ Int VG_(pipe) ( Int fd[2] )
 
 Off64T VG_(lseek) ( Int fd, Off64T offset, Int whence )
 {
-#  if defined(VGO_linux) || defined(VGP_amd64_darwin) || defined(VGP_amd64_freebsd) || defined(VGP_arm64_freebsd)
+#  if defined(VGO_linux) || defined(VGP_amd64_darwin) || defined(VGP_amd64_freebsd) || defined(VGP_arm64_freebsd) \
+	|| defined(VGO_openedition)
 #  if defined(__NR__llseek)
    Off64T result;
    SysRes res = VG_(do_syscall5)(__NR__llseek, fd,
@@ -511,7 +515,7 @@ SysRes VG_(stat) ( const HChar* file_name, struct vg_stat* vgbuf )
    SysRes res;
    VG_(memset)(vgbuf, 0, sizeof(*vgbuf));
 
-#  if defined(VGO_linux)
+#  if defined(VGO_linux) || defined(VGO_openedition)
    /* On Linux, first try with statx. If that doesn't work out, fall back to
       the stat64 or vanilla version. */
    { struct vki_statx buf;
@@ -525,7 +529,7 @@ SysRes VG_(stat) ( const HChar* file_name, struct vg_stat* vgbuf )
      }
    }
 #  endif
-#  if defined(VGO_linux) || defined(VGO_darwin)
+#  if defined(VGO_linux) || defined(VGO_darwin) || defined(VGO_openedition)
    /* Try with stat64. This is the second candidate on Linux, and the first
       one on Darwin. If that doesn't work out, fall back to vanilla version.
     */
@@ -595,7 +599,7 @@ Int VG_(fstat) ( Int fd, struct vg_stat* vgbuf )
    SysRes res;
    VG_(memset)(vgbuf, 0, sizeof(*vgbuf));
 
-#  if defined(VGO_linux)
+#  if defined(VGO_linux) || defined(VGO_openedition)
    /* On Linux, first try with statx. If that doesn't work out, fall back to
       the fstat64 or vanilla version. */
    { struct vki_statx buf;
@@ -610,7 +614,7 @@ Int VG_(fstat) ( Int fd, struct vg_stat* vgbuf )
      }
    }
 #endif
-#  if defined(VGO_linux) || defined(VGO_darwin)
+#  if defined(VGO_linux) || defined(VGO_darwin) || defined(VGO_openedition)
    /* Try with fstat64. This is the second candidate on Linux, and the first
       one on Darwin. If that doesn't work out, fall back to vanilla version.
     */
@@ -701,7 +705,7 @@ Long VG_(fsize) ( Int fd )
 SysRes VG_(getxattr) ( const HChar* file_name, const HChar* attr_name, Addr attr_value, SizeT attr_value_len )
 {
    SysRes res;
-#if defined(VGO_linux)
+#if defined(VGO_linux) || defined(VGO_openedition)
    res = VG_(do_syscall4)(__NR_getxattr, (UWord)file_name, (UWord)attr_name,
                           attr_value, attr_value_len);
 #else
@@ -720,7 +724,7 @@ Bool VG_(is_dir) ( const HChar* f )
 
 SysRes VG_(dup) ( Int oldfd )
 {
-#  if defined(VGO_linux) || defined(VGO_darwin) || defined(VGO_freebsd)
+#  if defined(VGO_linux) || defined(VGO_darwin) || defined(VGO_freebsd) || defined(VGO_openedition)
    return VG_(do_syscall1)(__NR_dup, oldfd);
 #  elif defined(VGO_solaris)
    return VG_(do_syscall3)(__NR_fcntl, oldfd, F_DUPFD, 0);
@@ -742,7 +746,7 @@ SysRes VG_(dup2) ( Int oldfd, Int newfd )
       return VG_(mk_SysRes_Success)(newfd);
    }
    return VG_(do_syscall3)(__NR_dup3, oldfd, newfd, 0);
-#  elif defined(VGO_linux) || defined(VGO_darwin) || defined(VGO_freebsd)
+#  elif defined(VGO_linux) || defined(VGO_darwin) || defined(VGO_freebsd) || defined(VGO_openedition)
    return VG_(do_syscall2)(__NR_dup2, oldfd, newfd);
 #  elif defined(VGO_solaris)
    return VG_(do_syscall3)(__NR_fcntl, oldfd, F_DUP2FD, newfd);
@@ -754,7 +758,7 @@ SysRes VG_(dup2) ( Int oldfd, Int newfd )
 /* Returns -1 on error. */
 Int VG_(fcntl) ( Int fd, Int cmd, Addr arg )
 {
-#  if defined(VGO_linux) || defined(VGO_solaris) || defined(VGO_freebsd)
+#  if defined(VGO_linux) || defined(VGO_solaris) || defined(VGO_freebsd) || defined(VGO_openedition)
 #  if defined(VGP_nanomips_linux)
    SysRes res = VG_(do_syscall3)(__NR_fcntl64, fd, cmd, arg);
 #  else
@@ -781,7 +785,7 @@ Int VG_(rename) ( const HChar* old_name, const HChar* new_name )
    SysRes res = VG_(do_syscall5)(__NR_renameat2, VKI_AT_FDCWD, (UWord)old_name,
                                  VKI_AT_FDCWD, (UWord)new_name, 0);
 
-#  elif defined(VGO_linux) || defined(VGO_darwin) || defined(VGO_freebsd)
+#  elif defined(VGO_linux) || defined(VGO_darwin) || defined(VGO_freebsd) || defined(VGO_openedition)
    SysRes res = VG_(do_syscall2)(__NR_rename, (UWord)old_name, (UWord)new_name);
 #  else
 #    error "Unknown OS"
@@ -794,7 +798,7 @@ Int VG_(unlink) ( const HChar* file_name )
 #  if defined(VGP_arm64_linux) || defined(VGP_nanomips_linux)
    SysRes res = VG_(do_syscall2)(__NR_unlinkat, VKI_AT_FDCWD,
                                                 (UWord)file_name);
-#  elif defined(VGO_linux) || defined(VGO_darwin) || defined(VGO_freebsd)
+#  elif defined(VGO_linux) || defined(VGO_darwin) || defined(VGO_freebsd) || defined(VGO_openedition)
    SysRes res = VG_(do_syscall1)(__NR_unlink, (UWord)file_name);
 #  elif defined(VGO_solaris)
    SysRes res = VG_(do_syscall3)(__NR_unlinkat, VKI_AT_FDCWD,
@@ -816,7 +820,7 @@ static HChar *startup_wd;
    changes. */
 void VG_(record_startup_wd) ( void )
 {
-#  if defined(VGO_linux) || defined(VGO_solaris) || defined(VGO_freebsd)
+#  if defined(VGO_linux) || defined(VGO_solaris) || defined(VGO_freebsd) || defined(VGO_openedition)
    /* Simple: just ask the kernel */
    SysRes res;
    SizeT szB = 0;
@@ -824,7 +828,7 @@ void VG_(record_startup_wd) ( void )
       szB += 500;
       startup_wd = VG_(realloc)("startup_wd", startup_wd, szB);
       VG_(memset)(startup_wd, 0, szB);
-#   if defined(VGO_linux) || defined(VGO_solaris)
+#   if defined(VGO_linux) || defined(VGO_solaris) || defined(VGO_openedition)
       res = VG_(do_syscall2)(__NR_getcwd, (UWord)startup_wd, szB-1);
 #   elif defined(VGO_freebsd)
       res = VG_(do_syscall2)(__NR___getcwd, (UWord)startup_wd, szB-1);
@@ -881,7 +885,7 @@ SysRes VG_(poll) (struct vki_pollfd *fds, Int nfds, Int timeout)
                           (UWord)fds, nfds, 
                           (UWord)(timeout >= 0 ? &timeout_ts : NULL),
                           (UWord)NULL);
-#  elif defined(VGO_linux)
+#  elif defined(VGO_linux) || defined(VGO_openedition)
    res = VG_(do_syscall3)(__NR_poll, (UWord)fds, nfds, timeout);
 #  elif defined(VGO_freebsd)
    res = VG_(do_syscall3)(__NR_poll, (UWord)fds, nfds, timeout);
@@ -918,7 +922,7 @@ SSizeT VG_(readlink) (const HChar* path, HChar* buf, SizeT bufsiz)
 #  if defined(VGP_arm64_linux) || defined(VGP_nanomips_linux)
    res = VG_(do_syscall4)(__NR_readlinkat, VKI_AT_FDCWD,
                                            (UWord)path, (UWord)buf, bufsiz);
-#  elif defined(VGO_linux) || defined(VGO_darwin) || defined(VGO_freebsd)
+#  elif defined(VGO_linux) || defined(VGO_darwin) || defined(VGO_freebsd) || defined(VGO_openedition)
    res = VG_(do_syscall3)(__NR_readlink, (UWord)path, (UWord)buf, bufsiz);
 #  elif defined(VGO_solaris)
    res = VG_(do_syscall4)(__NR_readlinkat, VKI_AT_FDCWD, (UWord)path,
@@ -929,7 +933,7 @@ SSizeT VG_(readlink) (const HChar* path, HChar* buf, SizeT bufsiz)
    return sr_isError(res) ? -1 : sr_Res(res);
 }
 
-#if defined(VGO_linux) || defined(VGO_solaris)
+#if defined(VGO_linux) || defined(VGO_solaris) || defined(VGO_openedition)
 Int VG_(getdents64) (Int fd, struct vki_dirent64 *dirp, UInt count)
 {
    SysRes res;
@@ -981,7 +985,7 @@ Int VG_(getdents64) (Int fd, struct vki_dirent64 *dirp, UInt count)
    nonzero otherwise. */
 Int VG_(access) ( const HChar* path, Bool irusr, Bool iwusr, Bool ixusr )
 {
-#  if defined(VGO_linux)
+#  if defined(VGO_linux) || defined(VGO_openedition)
    /* Very annoyingly, I cannot find any definition for R_OK et al in
       the kernel interfaces.  Therefore I reluctantly resort to
       hardwiring in these magic numbers that I determined by
@@ -996,7 +1000,7 @@ Int VG_(access) ( const HChar* path, Bool irusr, Bool iwusr, Bool ixusr )
              | (ixusr ? VKI_X_OK : 0);
 #  if defined(VGP_arm64_linux) || defined(VGP_nanomips_linux)
    SysRes res = VG_(do_syscall3)(__NR_faccessat, VKI_AT_FDCWD, (UWord)path, w);
-#  elif defined(VGO_linux) || defined(VGO_darwin) || defined(VGO_freebsd)
+#  elif defined(VGO_linux) || defined(VGO_darwin) || defined(VGO_freebsd) || defined(VGO_openedition)
    SysRes res = VG_(do_syscall2)(__NR_access, (UWord)path, w);
 #  elif defined(VGO_solaris)
    SysRes res = VG_(do_syscall4)(__NR_faccessat, VKI_AT_FDCWD, (UWord)path,
@@ -1006,7 +1010,7 @@ Int VG_(access) ( const HChar* path, Bool irusr, Bool iwusr, Bool ixusr )
 #  endif
    return sr_isError(res) ? 1 : 0;   
 
-#  if defined(VGO_linux)
+#  if defined(VGO_linux) || defined(VGO_openedition)
 #  undef VKI_R_OK
 #  undef VKI_W_OK
 #  undef VKI_X_OK
@@ -1110,7 +1114,7 @@ SysRes VG_(pread) ( Int fd, void* buf, Int count, OffT offset )
    SysRes res;
    // on 32 bits platforms, we receive a 32 bits OffT but
    // we must extend it to pass a long long 64 bits.
-#  if defined(VGP_x86_linux)
+#  if defined(VGP_x86_linux) || defined(VGP_s390x_openedition)
    vg_assert(sizeof(OffT) == 4);
    res = VG_(do_syscall5)(__NR_pread64, fd, (UWord)buf, count, 
                           offset, 0); // Little endian long long
@@ -1126,7 +1130,7 @@ SysRes VG_(pread) ( Int fd, void* buf, Int count, OffT offset )
                           0, // Padding needed on PPC32
                           0, offset); // Big endian long long
    return res;
-#  elif (defined(VGP_mips32_linux) || defined(VGP_nanomips_linux)) \
+#  elif (defined(VGP_mips32_linux) || defined(VGP_nanomips_linux))\
      && (VKI_LITTLE_ENDIAN)
    vg_assert(sizeof(OffT) == 4);
    res = VG_(do_syscall6)(__NR_pread64, fd, (UWord)buf, count, 
@@ -1303,7 +1307,8 @@ UShort VG_(ntohs) ( UShort x )
 */
 Int VG_(connect_via_socket)( const HChar* str )
 {
-#  if defined(VGO_linux) || defined(VGO_darwin) || defined(VGO_solaris) || defined(VGO_freebsd)
+#  if defined(VGO_linux) || defined(VGO_darwin) || defined(VGO_solaris) || defined(VGO_freebsd) \
+	|| defined(VGO_openedition)
    Int sd, res;
    struct vki_sockaddr_in servAddr;
    UInt   ip   = 0;
@@ -1394,7 +1399,7 @@ Int VG_(socket) ( Int domain, Int type, Int protocol )
 {
 #  if defined(VGP_x86_linux) || defined(VGP_ppc32_linux) \
       || defined(VGP_ppc64be_linux) || defined(VGP_ppc64le_linux) \
-      || defined(VGP_s390x_linux)
+      || defined(VGP_s390x_linux) || defined(VGP_s390x_openedition)
    SysRes res;
    UWord  args[3];
    args[0] = domain;
@@ -1449,7 +1454,7 @@ Int my_connect ( Int sockfd, struct vki_sockaddr_in* serv_addr, Int addrlen )
 {
 #  if defined(VGP_x86_linux) || defined(VGP_ppc32_linux) \
       || defined(VGP_ppc64be_linux) || defined(VGP_ppc64le_linux) \
-      || defined(VGP_s390x_linux)
+      || defined(VGP_s390x_linux) || defined(VGP_s390x_openedition)
    SysRes res;
    UWord  args[3];
    args[0] = sockfd;
@@ -1495,7 +1500,7 @@ Int VG_(write_socket)( Int sd, const void *msg, Int count )
 
 #  if defined(VGP_x86_linux) || defined(VGP_ppc32_linux) \
       || defined(VGP_ppc64be_linux) || defined(VGP_ppc64le_linux) \
-      || defined(VGP_s390x_linux)
+      || defined(VGP_s390x_linux) || defined(VGP_s390x_openedition)
    SysRes res;
    UWord  args[4];
    args[0] = sd;
@@ -1533,7 +1538,7 @@ Int VG_(getsockname) ( Int sd, struct vki_sockaddr *name, Int *namelen)
 #  if defined(VGP_x86_linux) || defined(VGP_ppc32_linux) \
       || defined(VGP_ppc64be_linux) || defined(VGP_ppc64le_linux) \
       || defined(VGP_s390x_linux) \
-      || defined(VGP_mips32_linux)
+      || defined(VGP_mips32_linux) || defined(VGP_s390x_openedition)
    SysRes res;
    UWord  args[3];
    args[0] = sd;
@@ -1573,7 +1578,7 @@ Int VG_(getpeername) ( Int sd, struct vki_sockaddr *name, Int *namelen)
 #  if defined(VGP_x86_linux) || defined(VGP_ppc32_linux) \
       || defined(VGP_ppc64be_linux) || defined(VGP_ppc64le_linux) \
       || defined(VGP_s390x_linux) \
-      || defined(VGP_mips32_linux)
+      || defined(VGP_mips32_linux) || defined(VGP_s390x_openedition)
    SysRes res;
    UWord  args[3];
    args[0] = sd;
@@ -1612,7 +1617,7 @@ Int VG_(getsockopt) ( Int sd, Int level, Int optname, void *optval,
 {
 #  if defined(VGP_x86_linux) || defined(VGP_ppc32_linux) \
       || defined(VGP_ppc64be_linux) || defined(VGP_ppc64le_linux) \
-      || defined(VGP_s390x_linux)
+      || defined(VGP_s390x_linux) || defined(VGP_s390x_openedition)
    SysRes res;
    UWord  args[5];
    args[0] = sd;
@@ -1657,7 +1662,7 @@ Int VG_(setsockopt) ( Int sd, Int level, Int optname, void *optval,
 {
 #  if defined(VGP_x86_linux) || defined(VGP_ppc32_linux) \
       || defined(VGP_ppc64be_linux) || defined(VGP_ppc64le_linux) \
-      || defined(VGP_s390x_linux)
+      || defined(VGP_s390x_linux) || defined(VGP_s390x_openedition)
    SysRes res;
    UWord  args[5];
    args[0] = sd;
