diff --git a/coregrind/m_signals.c b/coregrind/m_signals.c
index 66fbbfe72..3b525c261 100644
--- a/coregrind/m_signals.c
+++ b/coregrind/m_signals.c
@@ -574,6 +574,30 @@ VgHashTable *ht_sigchld_ignore = NULL;
       VG_(mk_SysRes_s390x_linux)((uc)->uc_mcontext.regs.gprs[2])
 #  define VG_UCONTEXT_LINK_REG(uc) ((uc)->uc_mcontext.regs.gprs[14])
 
+#  define VG_UCONTEXT_TO_UnwindStartRegs(srP, uc)        \
+      { (srP)->r_pc = (ULong)((uc)->uc_mcontext.regs.psw.addr);    \
+        (srP)->r_sp = (ULong)((uc)->uc_mcontext.regs.gprs[15]);    \
+        (srP)->misc.S390X.r_fp = (uc)->uc_mcontext.regs.gprs[11];  \
+        (srP)->misc.S390X.r_lr = (uc)->uc_mcontext.regs.gprs[14];  \
+        (srP)->misc.S390X.r_f0 = (uc)->uc_mcontext.fpregs.fprs[0]; \
+        (srP)->misc.S390X.r_f1 = (uc)->uc_mcontext.fpregs.fprs[1]; \
+        (srP)->misc.S390X.r_f2 = (uc)->uc_mcontext.fpregs.fprs[2]; \
+        (srP)->misc.S390X.r_f3 = (uc)->uc_mcontext.fpregs.fprs[3]; \
+        (srP)->misc.S390X.r_f4 = (uc)->uc_mcontext.fpregs.fprs[4]; \
+        (srP)->misc.S390X.r_f5 = (uc)->uc_mcontext.fpregs.fprs[5]; \
+        (srP)->misc.S390X.r_f6 = (uc)->uc_mcontext.fpregs.fprs[6]; \
+        (srP)->misc.S390X.r_f7 = (uc)->uc_mcontext.fpregs.fprs[7]; \
+      }
+
+#elif defined(VGP_s390x_openedition)
+
+#  define VG_UCONTEXT_INSTR_PTR(uc)       ((uc)->uc_mcontext.regs.psw.addr)
+#  define VG_UCONTEXT_STACK_PTR(uc)       ((uc)->uc_mcontext.regs.gprs[15])
+#  define VG_UCONTEXT_FRAME_PTR(uc)       ((uc)->uc_mcontext.regs.gprs[11])
+#  define VG_UCONTEXT_SYSCALL_SYSRES(uc)                        \
+      VG_(mk_SysRes_s390x_openedition)((uc)->uc_mcontext.regs.gprs[2])
+#  define VG_UCONTEXT_LINK_REG(uc) ((uc)->uc_mcontext.regs.gprs[14])
+
 #  define VG_UCONTEXT_TO_UnwindStartRegs(srP, uc)        \
       { (srP)->r_pc = (ULong)((uc)->uc_mcontext.regs.psw.addr);    \
         (srP)->r_sp = (ULong)((uc)->uc_mcontext.regs.gprs[15]);    \
@@ -679,7 +703,7 @@ VgHashTable *ht_sigchld_ignore = NULL;
 
 /* These macros allow use of uniform names when working with
    both the Linux and Darwin vki definitions. */
-#if defined(VGO_linux)
+#if defined(VGO_linux) || defined(VGO_openedition)
 #  define VKI_SIGINFO_si_addr  _sifields._sigfault._addr
 #  define VKI_SIGINFO_si_pid   _sifields._kill._pid
 #elif defined(VGO_darwin) || defined(VGO_solaris) || defined(VGO_freebsd)
@@ -1044,6 +1068,14 @@ extern void my_sigreturn(void);
    " svc " #name "\n" \
    ".previous\n"
 
+#elif defined(VGP_s390x_openedition)
+#  define _MY_SIGRETURN(name) \
+   ".text\n" \
+   ".globl my_sigreturn\n" \
+   "my_sigreturn:\n" \
+   " svc " #name "\n" \
+   ".previous\n"
+
 #elif defined(VGP_mips32_linux)
 #  define _MY_SIGRETURN(name) \
    ".text\n" \
@@ -1706,7 +1738,7 @@ void VG_(kill_self)(Int sigNo)
 // pass in some other details that can help when si_code is unreliable.
 static Bool is_signal_from_kernel(ThreadId tid, int signum, int si_code)
 {
-#  if defined(VGO_linux) || defined(VGO_solaris)
+#  if defined(VGO_linux) || defined(VGO_solaris) || defined(VGO_openedition)
    // On Linux, SI_USER is zero, negative values are from the user, positive
    // values are from the kernel.  There are SI_FROMUSER and SI_FROMKERNEL
    // macros but we don't use them here because other platforms don't have
@@ -1838,7 +1870,7 @@ static void default_action(const vki_siginfo_t *info, ThreadId tid)
    if (!terminate)
       return;			/* nothing to do */
 
-#if defined(VGO_linux)
+#if defined(VGO_linux) || defined(VGO_openedition)
    if (terminate && (tst->ptrace & VKI_PT_PTRACED)
        && (sigNo != VKI_SIGKILL)) {
       VG_(kill)(VG_(getpid)(), VKI_SIGSTOP);
@@ -1978,7 +2010,7 @@ static void default_action(const vki_siginfo_t *info, ThreadId tid)
          likely cause a segfault. */
       if (VG_(is_valid_tid)(tid)) {
          Word first_ip_delta = 0;
-#if defined(VGO_linux) || defined(VGO_solaris)
+#if defined(VGO_linux) || defined(VGO_solaris) || defined(VGO_openedition)
          /* Make sure that the address stored in the stack pointer is 
             located in a mapped page. That is not necessarily so. E.g.
             consider the scenario where the stack pointer was decreased
@@ -2085,7 +2117,7 @@ static void deliver_signal ( ThreadId tid, const vki_siginfo_t *info,
    void			*handler_fn;
    ThreadState		*tst = VG_(get_ThreadState)(tid);
 
-#if defined(VGO_linux)
+#if defined(VGO_linux) || defined(VGO_openedition)
    /* If this signal is SIGCHLD and it came from a process which valgrind
       created for some internal use, then it should not be delivered to
       the client.  */
@@ -2430,7 +2462,7 @@ static vki_siginfo_t *next_queued(ThreadId tid, const vki_sigset_t *set)
 
 static int sanitize_si_code(int si_code)
 {
-#if defined(VGO_linux)
+#if defined(VGO_linux) || defined(VGO_openedition)
    /* The linux kernel uses the top 16 bits of si_code for it's own
       use and only exports the bottom 16 bits to user space - at least
       that is the theory, but it turns out that there are some kernels
@@ -2772,7 +2804,7 @@ void sync_signalhandler_from_user ( ThreadId tid,
       if (VG_(clo_trace_signals))
          VG_(dmsg)("Routing user-sent sync signal %d via queue\n", sigNo);
 
-#     if defined(VGO_linux)
+#     if defined(VGO_linux) || defined(VGO_openedition)
       /* On Linux, first we have to do a sanity check of the siginfo. */
       if (info->VKI_SIGINFO_si_pid == 0) {
          /* There's a per-user limit of pending siginfo signals.  If
@@ -2809,7 +2841,7 @@ void sync_signalhandler_from_user ( ThreadId tid,
 #     endif
 
       qtid = 0;         /* shared pending by default */
-#     if defined(VGO_linux)
+#     if defined(VGO_linux) || defined(VGO_openedition)
       if (info->si_code == VKI_SI_TKILL)
          qtid = tid;    /* directed to us specifically */
 #     endif
