diff --git a/coregrind/pub_core_initimg.h b/coregrind/pub_core_initimg.h
index cfbd41f8b..b581aa4a0 100644
--- a/coregrind/pub_core_initimg.h
+++ b/coregrind/pub_core_initimg.h
@@ -66,7 +66,7 @@ void VG_(ii_finalise_image)( IIFinaliseImageInfo );
 
 /* ------------------------- Linux ------------------------- */
 
-#if defined(VGO_linux) || defined(VGO_freebsd)
+#if defined(VGO_linux) || defined(VGO_freebsd) || defined(VGO_openedition)
 
 struct _IICreateImageInfo {
    /* ------ Mandatory fields ------ */
