diff --git a/coregrind/pub_core_machine.h b/coregrind/pub_core_machine.h
index a1afbe61c..c25e06362 100644
--- a/coregrind/pub_core_machine.h
+++ b/coregrind/pub_core_machine.h
@@ -79,7 +79,7 @@
 #  undef  VG_ELF_MACHINE
 #  undef  VG_ELF_CLASS
 #  undef  VG_PLAT_USES_PPCTOC
-#elif defined(VGP_s390x_linux)
+#elif defined(VGP_s390x_linux) || defined(VGP_s390x_openedition)
 #  define VG_ELF_DATA2XXX     ELFDATA2MSB
 #  define VG_ELF_MACHINE      EM_S390
 #  define VG_ELF_CLASS        ELFCLASS64
