diff --git a/include/pub_tool_libcsetjmp.h b/include/pub_tool_libcsetjmp.h
index e9638c3d0..b8b99aed0 100644
--- a/include/pub_tool_libcsetjmp.h
+++ b/include/pub_tool_libcsetjmp.h
@@ -134,6 +134,14 @@ UWord VG_MINIMAL_SETJMP(VG_MINIMAL_JMP_BUF(_env));
 __attribute__((noreturn))
 void  VG_MINIMAL_LONGJMP(VG_MINIMAL_JMP_BUF(_env));
 
+#elif defined(VGP_s390x_openedition)
+
+#define VG_MINIMAL_JMP_BUF(_name)        ULong _name [10 + 8]
+__attribute__((returns_twice))
+UWord VG_MINIMAL_SETJMP(VG_MINIMAL_JMP_BUF(_env));
+__attribute__((noreturn))
+void  VG_MINIMAL_LONGJMP(VG_MINIMAL_JMP_BUF(_env));
+
 #elif defined(__clang__) && defined(VGP_arm64_linux)
 
 #define VG_MINIMAL_JMP_BUF(_name)        UWord _name [13]
