diff --git a/VEX/priv/s390_disasm.h b/VEX/priv/s390_disasm.h
index d6b2fbc8e..eec41f8ac 100644
--- a/VEX/priv/s390_disasm.h
+++ b/VEX/priv/s390_disasm.h
@@ -80,8 +80,19 @@ enum {
    S390_XMNM_BC = 2,
    S390_XMNM_BRC = 3,
    S390_XMNM_BRCL = 4,
-   S390_XMNM_CLS = 5,
-   S390_XMNM_BIC = 6,
+   S390_XMNM_LOCR = 5,
+   S390_XMNM_LOCGR = 6,
+   S390_XMNM_LOC = 7,
+   S390_XMNM_LOCG = 8,
+   S390_XMNM_STOC = 9,
+   S390_XMNM_STOCG = 10,
+   S390_XMNM_STOCFH = 11,
+   S390_XMNM_LOCFH = 12,
+   S390_XMNM_LOCFHR = 13,
+   S390_XMNM_LOCHI = 14,
+   S390_XMNM_LOCGHI = 15,
+   S390_XMNM_LOCHHI = 16,
+   S390_XMNM_BIC = 17
 };
 
 void s390_disasm(UInt command, ...);
