diff --git a/coregrind/vg_preloaded.c b/coregrind/vg_preloaded.c
index 5bec51d76..460c7d7b1 100644
--- a/coregrind/vg_preloaded.c
+++ b/coregrind/vg_preloaded.c
@@ -49,7 +49,7 @@
 #include <features.h>
 #endif
 
-#if !defined(VGO_darwin)
+#if !defined(VGO_darwin) && !defined(VGO_openedition)
 /* Instruct GDB via a .debug_gdb_scripts section to load the valgrind and tool
    front-end commands.  */
 /* Note: The "MS" section flags are to remove duplicates.  */
@@ -66,7 +66,8 @@ DEFINE_GDB_PY_SCRIPT(VG_GDBSCRIPTS_DIR "/valgrind-monitor.py")
 #endif
 #endif
 
-#if defined(VGO_linux) || defined(VGO_solaris) || defined(VGO_freebsd)
+#if defined(VGO_linux) || defined(VGO_solaris) || defined(VGO_freebsd) \
+    || defined(VGO_openedition)
 
 /* ---------------------------------------------------------------------
    Hook for running __gnu_cxx::__freeres() and __libc_freeres() once
@@ -80,7 +81,8 @@ void VG_NOTIFY_ON_LOAD(freeres)(Vg_FreeresToRun to_run)
       && !defined(VGPV_arm_linux_android) \
       && !defined(VGPV_x86_linux_android) \
       && !defined(VGPV_mips32_linux_android) \
-      && !defined(VGPV_arm64_linux_android)
+      && !defined(VGPV_arm64_linux_android) \
+      && !defined(VGO_openedition)	
 
    /* g++ mangled __gnu_cxx::__freeres yields -> _ZN9__gnu_cxx9__freeresEv */
    extern void _ZN9__gnu_cxx9__freeresEv(void) __attribute__((weak));
@@ -95,7 +97,8 @@ void VG_NOTIFY_ON_LOAD(freeres)(Vg_FreeresToRun to_run)
       && !defined(VGPV_arm_linux_android) \
       && !defined(VGPV_x86_linux_android) \
       && !defined(VGPV_mips32_linux_android) \
-      && !defined(VGPV_arm64_linux_android)
+      && !defined(VGPV_arm64_linux_android) \
+      && !defined(VGO_openedition)
 
    extern void __libc_freeres(void) __attribute__((weak));
    if (((to_run & VG_RUN__LIBC_FREERES) != 0) &&
@@ -416,6 +419,10 @@ out:
    return pid;
 }
 
+#elif defined(VGO_openedition)
+
+/* No z/OS specific preloaded redirections yet. */
+
 #else
 #  error Unknown OS
 #endif
